---
title: "`r params$title`"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: true
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
params:
  title: null
  cache_path: null
  rerun: false
---

```{r setup, include = FALSE}
knitr::opts_knit$set(root.dir = "../..")
knitr::opts_chunk$set(
  warning = FALSE,
  message = FALSE,
  dpi = 300,
  cache.path = params$cache_path,
  fig.align = "center"
)
```

```{r setup2, include = FALSE}
# libraries
library(magrittr)
library(dplyr)
library(ggplot2)
library(patchwork)
library(biomaRt)

# set ggplot presets
theme_set(theme_classic())

# function: plot phased baf
plot_baf <- function(p_dat, min_depth = 5) {
  # prep data
  p_dat2 <-
    p_dat %>%
    filter(total_depth >= min_depth) %>%
    dplyr::mutate(mut_baf = 1 - mut_vaf,
                  chr = factor(sub("^chr", "", chr),
                               levels = c(as.character(1:22), "X", "Y")),
                  `pos (Mb)` = pos / 1e6) %>%
    tidyr::pivot_longer(cols = c("mut_vaf", "mut_baf"), names_to = "vaf_type")

  # prep phased data
  p_dat_phased <-
    p_dat2 %>%
    filter(!is.na(hap_A_vaf) & vaf_type == "mut_vaf")

  # plot
  p_dat2 %>%
    ggplot(aes(x = `pos (Mb)`, y = value)) +
    # add line at vaf = 0.5
    geom_hline(yintercept = 0.5, colour = "red") +
    # add baf points
    geom_point(size = 0.8, alpha = 0.2) +
    # add phased points
    geom_point(data = p_dat_phased, aes(y = hap_A_vaf), colour = "#e03ba9",
               size = 1) +
    geom_point(data = p_dat_phased, aes(y = hap_B_vaf), colour = "#00ccff",
               size = 1) +
    # increase x axis breaks
    scale_x_continuous(breaks = pretty(p_dat2$`pos (Mb)`, n = 20)) +
    scale_y_continuous(expand = c(0.01, 0.01)) +
    theme_classic() +
    theme(panel.grid.major = element_line(),
          panel.grid.minor.x = element_line(),
          strip.background = element_rect(color = "grey", fill = NA,
                                          linewidth = 0, linetype = "solid")) +
    # rotate x axis text using guides
    guides(x = guide_axis(angle = -90))
}
```

# Parameters

```{r params, echo = FALSE}
params
```

# Load SNPs

```{r load_snps, eval = F}
snps <-
  c("dna", "dnahyb") %>%
  purrr::set_names() %>%
  purrr::map(function(type) {
    file.path("out/resolveome/nf-resolveome", type,
              "PD63118/genotyping/snps/PD63118_genotyped_snps.tsv") %>%
      readr::read_tsv() %>%
      dplyr::filter(chr == "1", pos < 5e6) %>%
      dplyr::mutate(cell_id = sub("_dna.*", "", id)) %>%
      split(.$cell_id)
  })

saveRDS(snps, "reports/20250827_plot_phased_1p_5Mb/snps.rds")
```

```{r reload_snps}
snps <- readRDS("reports/20250827_plot_phased_1p_5Mb/snps.rds")
```

# Phase SNPs

```{r phase_snps}
# load phasing from plate3_wellA2
plate3_wellA2 <-
  readRDS("out/phase_snps/phase_snps_cache/plate3_wellA2_phased_snps.rds")

# phase snps
cell_ids <- unique(unlist(lapply(snps, names)))
phased_snps <- list()
purrr::walk(cell_ids, function(cell_id_i) {
  phased_snps[[cell_id_i]] <<- list()
  purrr::walk(c("dna", "dnahyb"), function(type_i) {
    if (!is.null(snps[[type_i]][[cell_id_i]])) {
      phased_snps[[cell_id_i]][[type_i]] <<-
        snps[[type_i]][[cell_id_i]] %>%
        left_join(
          plate3_wellA2 %>%
            distinct(chr, pos, ref, alt, hap_A_type, hap_B_type) %>%
            filter(!is.na(hap_A_type)),
          by = c("chr", "pos", "ref", "alt")) %>%
        mutate(hap_A_vaf = case_when(hap_A_type == "alt" ~ mut_vaf,
                                     hap_A_type == "ref" ~ 1 - mut_vaf),
               hap_B_vaf = case_when(hap_B_type == "alt" ~ mut_vaf,
                                     hap_B_type == "ref" ~ 1 - mut_vaf))
    }
  })
})
```

# Plot BAFs

```{r plot_bafs, fig.width = 10, eval = F}
purrr::walk2(names(phased_snps), phased_snps, function(cell_id_i, x) {

  print(cell_id_i)

  purrr::walk2(names(x), x, function(type_i, i) {

    print(type_i)

    min_depth <- ifelse(type_i == "dna", 5, 10)

    # first 5Mb
    p <-
      plot_baf(i, min_depth = min_depth) +
      ggtitle(paste0(cell_id_i,  " - ", type_i, " (first 5Mb, min depth = ",
                     min_depth, ")"))
    print(p)

    # 200kb around TNFRSF14
    p <-
      i %>%
      filter(pos > 2487078 - 2e5, pos < 2487078 + 2e5) %>%
      plot_baf(min_depth = min_depth) +
      ggtitle(paste0(cell_id_i, " - ", type_i,
                     " (200kb around TNFRSF14, min depth = ", min_depth, ")"))
    print(p)

  })
})
```

# Plot BAFs in cells of interest

```{r coi, fig.width = 10}
# cells of interest
cois <- c("plate10_wellF3", "plate3_wellD9", "plate10_wellF8")

# load tnfrsf14 mutations
muts <-
  c("dna", "dnahyb") %>%
  purrr::set_names() %>%
  purrr::map(function(type) {
    file.path("out/resolveome/nf-resolveome", type,
              "PD63118/genotyping/mutations",
              "PD63118_annotated_mutations.tsv") %>%
      readr::read_tsv() %>%
      dplyr::mutate(cell_id = sub("_dna.*", "", id)) %>%
      dplyr::filter(gene == "TNFRSF14", cell_id %in% cois)
  }) %>%
  bind_rows(.id = "seq_type") %>%
  group_by(cell_id, gene, aachange, chr, pos, ref, alt) %>%
  summarise(total_depth = sum(total_depth),
            alt_depth = sum(alt_depth),
            .groups = "drop") %>%
  mutate(alt_vaf = alt_depth / total_depth) %>%
  split(.$cell_id)

# get tnfrsf14 exons and introns
ensembl38 <-
  useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl", GRCh = 38)
ct <-
  getBM(attributes = c("ensembl_transcript_id", "transcript_is_canonical"),
        filters = "hgnc_symbol", values = "TNFRSF14", mart = ensembl38) %>%
  filter(transcript_is_canonical == 1) %>%
  pull(ensembl_transcript_id)
ensembl37 <-
  useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl", GRCh = 37)
exons <-
  getBM(attributes = c("chromosome_name", "exon_chrom_start", "exon_chrom_end",
                       "strand", "ensembl_exon_id", "ensembl_transcript_id"),
        filters = "hgnc_symbol", values = "TNFRSF14", mart = ensembl37) %>%
  filter(ensembl_transcript_id %in% ct) %>%
  group_by(ensembl_transcript_id) %>%
  mutate(start = min(exon_chrom_start),
         end = max(exon_chrom_end)) %>%
  ungroup() %>%
  mutate(start_mb = start / 1e6, end_mb = end / 1e6,
         exon_chrom_start_mb = exon_chrom_start / 1e6,
         exon_chrom_end_mb = exon_chrom_end / 1e6,
         gene = "TNFRSF14")

# plots bafs and transcripts
purrr::walk(cois, function(cell_id_i) {

  print(cell_id_i)

  # get mutations
  muts_i <-
    muts[[cell_id_i]] %>%
    filter(alt_vaf > 0.2, alt_depth > 1)

  purrr::walk(names(phased_snps[[cell_id_i]]), function(type_i) {

    print(type_i)

    # get snps
    phased_snps_i <- phased_snps[[cell_id_i]][[type_i]]

    for (min_depth in c(0, ifelse(type_i == "dna", 5, 10))) {
      for (range in c(5e3, 1e5, 2e5)) {

        # define x limits
        x_limits <- c(min(exons$start) - range, max(exons$end) + range) / 1e6

        # prep data
        p_dat <-
          phased_snps_i %>%
          filter(total_depth >= min_depth,
                 pos > min(exons$start) - range,
                 pos < max(exons$end) + range) %>%
          dplyr::mutate(mut_baf = 1 - mut_vaf,
                        chr = factor(sub("^chr", "", chr),
                                     levels = c(as.character(1:22), "X", "Y")),
                        `pos (Mb)` = pos / 1e6) %>%
          tidyr::pivot_longer(cols = c("mut_vaf", "mut_baf"),
                              names_to = "vaf_type") %>%
          tidyr::pivot_longer(cols = c("hap_A_vaf", "hap_B_vaf"),
                              names_to = "haplotype", values_to = "hap_vaf") %>%
          mutate(haplotype = case_when(is.na(hap_vaf) ~ "unphased",
                                       TRUE ~ haplotype))

        # plot bafs
        p_baf <-
          p_dat %>%
          ggplot(aes(x = `pos (Mb)`, y = value)) +
          # add box around TNFRSF14 position
          geom_rect(data = exons, inherit.aes = FALSE,
                    aes(xmin = start_mb, xmax = end_mb,
                        ymin = -Inf, ymax = Inf),
                    fill = "#f2f2f2") +
          # add line at vaf = 0.5
          geom_hline(yintercept = 0.5, colour = "red") +
          # add baf points
          geom_point(size = 0.8, alpha = 0.2) +
          # add phased points
          geom_point(aes(y = hap_vaf, colour = haplotype, alpha = haplotype),
                     size = 1) +
          scale_colour_manual(values = c(hap_A_vaf = "#e03ba9",
                                         hap_B_vaf = "#00ccff",
                                         unphased = "grey")) +
          scale_alpha_manual(values = c(hap_A_vaf = 1,
                                        hap_B_vaf = 1,
                                        unphased = 0.2)) +
          # axes
          xlim(x_limits) +
          scale_y_continuous(expand = c(0.01, 0.01)) +
          theme_classic() +
          theme(panel.grid.major = element_blank(),
                panel.grid.minor = element_blank(),
                legend.position = "none") +
          labs(y = "VAF")

        # exons
        p_transcript <-
          exons %>%
          ggplot(aes(x = exon_chrom_start_mb, xend = exon_chrom_end_mb,
                     y = gene, yend = gene)) +
          geom_segment(linewidth = 5) +
          geom_segment(aes(x = start_mb, xend = end_mb), linewidth = 1) +
          geom_text(aes(x = end_mb, label = gene),
                    hjust = 0, nudge_x = diff(x_limits) * 0.01) +
          ggrepel::geom_label_repel(
            data = muts_i,
            aes(x = pos / 1e6, y = gene,
                label = paste0(aachange, " (", round(alt_vaf, 2),
                               ", ", alt_depth, "/", total_depth, ")")),
                inherit.aes = FALSE, nudge_y = 0.5,
                min.segment.length = 0, direction = "x") +
          geom_point(data = muts_i,
                     aes(x = pos / 1e6, y = gene),
                     inherit.aes = FALSE,
                     shape = 21, fill = "red", size = 3) +
          labs(x = "pos (Mb)") +
          xlim(x_limits) +
          theme_void() +
          theme(plot.title = element_text()) +
          ggtitle(paste0(cell_id_i, " - ", type_i, " (", range / 1e3,
                         "kb around TNFRSF14, min depth = ", min_depth, ")"))


        # plot
        p <-
          (p_transcript / plot_spacer() / p_baf) + 
          plot_layout(heights = c(1, -0.5, 4))
        print(p)
      }
    }
  })
})
```

# Plot DNA+DNAhyb BAFs in cells of interest

```{r coi_dna_dnahyb, fig.width = 10}
# plots bafs and transcripts
purrr::walk(cois, function(cell_id_i) {

  print(cell_id_i)

  # get mutations
  muts_i <-
    muts[[cell_id_i]] %>%
    filter(alt_vaf > 0.2, alt_depth > 1)

  # combine reads from dna + dnahyb
  phased_snps_i <-
    phased_snps[[cell_id_i]] %>%
    bind_rows(.id = "seq_type") %>%
    group_by(chr, pos, ref, alt, cell_id, hap_A_type, hap_B_type, hap_A_vaf, hap_B_vaf) %>%
    summarise(mut_depth = sum(mut_depth),
              total_depth = sum(total_depth),
              .groups = "drop") %>%
    mutate(mut_vaf = mut_depth / total_depth)

  for (min_depth in c(0, 10)) {
    for (range in c(5e3, 1e5, 2e5)) {

      # define x limits
      x_limits <- c(min(exons$start) - range, max(exons$end) + range) / 1e6

      # prep data
      p_dat <-
        phased_snps_i %>%
        filter(total_depth >= min_depth,
               pos > min(exons$start) - range,
               pos < max(exons$end) + range) %>%
        dplyr::mutate(mut_baf = 1 - mut_vaf,
                      chr = factor(sub("^chr", "", chr),
                                    levels = c(as.character(1:22), "X", "Y")),
                      `pos (Mb)` = pos / 1e6) %>%
        tidyr::pivot_longer(cols = c("mut_vaf", "mut_baf"),
                            names_to = "vaf_type") %>%
        tidyr::pivot_longer(cols = c("hap_A_vaf", "hap_B_vaf"),
                            names_to = "haplotype", values_to = "hap_vaf") %>%
        mutate(haplotype = case_when(is.na(hap_vaf) ~ "unphased",
                                     TRUE ~ haplotype))

      # plot bafs
      p_baf <-
        p_dat %>%
        ggplot(aes(x = `pos (Mb)`, y = value)) +
        # add box around TNFRSF14 position
        geom_rect(data = exons, inherit.aes = FALSE,
                  aes(xmin = start_mb, xmax = end_mb,
                      ymin = -Inf, ymax = Inf),
                  fill = "#f2f2f2") +
        # add line at vaf = 0.5
        geom_hline(yintercept = 0.5, colour = "red") +
        # add baf points
        geom_point(size = 0.8, alpha = 0.2) +
        # add phased points
        geom_point(aes(y = hap_vaf, colour = haplotype, alpha = haplotype),
                    size = 1) +
        scale_colour_manual(values = c(hap_A_vaf = "#e03ba9",
                                        hap_B_vaf = "#00ccff",
                                        unphased = "grey")) +
        scale_alpha_manual(values = c(hap_A_vaf = 1,
                                      hap_B_vaf = 1,
                                      unphased = 0.2)) +
        # axes
        xlim(x_limits) +
        scale_y_continuous(expand = c(0.01, 0.01)) +
        theme_classic() +
        theme(panel.grid.major = element_blank(),
              panel.grid.minor = element_blank(),
              legend.position = "none") +
        labs(y = "VAF")

      # exons
      p_transcript <-
        exons %>%
        ggplot(aes(x = exon_chrom_start_mb, xend = exon_chrom_end_mb,
                    y = gene, yend = gene)) +
        geom_segment(linewidth = 5) +
        geom_segment(aes(x = start_mb, xend = end_mb), linewidth = 1) +
        geom_text(aes(x = end_mb, label = gene),
                  hjust = 0, nudge_x = diff(x_limits) * 0.01) +
        ggrepel::geom_label_repel(
          data = muts_i,
          aes(x = pos / 1e6, y = gene,
              label = paste0(aachange, " (", round(alt_vaf, 2),
                             ", ", alt_depth, "/", total_depth, ")")),
              inherit.aes = FALSE, nudge_y = 0.5,
              min.segment.length = 0, direction = "x") +
        geom_point(data = muts_i,
                   aes(x = pos / 1e6, y = gene),
                   inherit.aes = FALSE,
                   shape = 21, fill = "red", size = 3) +
        labs(x = "pos (Mb)") +
        xlim(x_limits) +
        theme_void() +
        theme(plot.title = element_text()) +
        ggtitle(paste0(cell_id_i, " - dna + dnahyb (", range / 1e3,
                       "kb around TNFRSF14, min depth = ", min_depth, ")"))


      # plot
      p <-
        (p_transcript / plot_spacer() / p_baf) +
        plot_layout(heights = c(1, -0.5, 4))
      print(p)
    }
  }
})
```