---
title: "Shared mutations from the bj-somatic-variantcalling DNA + targeted DNA output on filtered cells"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: true
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
params:
  rerun: false
---

```{r setup, include = F}
# rmarkdown::render('reports/20250331_shared_muts_from_bj-somatic-variantcalling_filter_cells.Rmd', output_file = '20250331_shared_muts_from_bj-somatic-variantcalling_filter_cells.html', output_dir = 'reports/')

# chunk options
knitr::opts_knit$set(root.dir = "../")
knitr::opts_chunk$set(warning = FALSE,
                      dpi = 600,
                      message = FALSE,
                      cache.path = "reports/20250331_shared_muts_from_bj-somatic-variantcalling_filter_cells_cache/")

# set buffer for reading long lines in vcfs
Sys.setenv("VROOM_CONNECTION_SIZE" = 2^20) # 1MB buffer

# libraries
library(magrittr)
library(ggplot2)
library(dndscv)
library(GenomicRanges)
library(ggtree)
library(dplyr)

# palettes
impact_colours <- c("Essential_Splice" = "#8e55a3",
                    "Nonsense" = "#682c8b",
                    "Missense" = "#5f9fa0",
                    "Indels" = "#d4773f",
                    "Synonymous" = "#b3b4b4",
                    "Non-coding" = "#606060")

# function: get the trinucleotide context of the muts
get_mut_trinucs <- function(dat) {
  dat %>%
    # get snps only
    filter(nchar(ref) == 1, nchar(alt) == 1) %>%
    # get the trinucleotide context
    rowwise() %>%
    mutate(
      trinuc_ref = Rsamtools::scanFa("data/gatk/grch38/genome.fa",
                                     GRanges(chr, IRanges(pos - 1,
                                                          pos + 1))) %>%
                    as.vector()) %>%
    ungroup() %>%
    # annotate the mutation from the pyrimidine base
    mutate(
      sub = paste(ref, alt, sep = ">"),
      sub_py = case_when(
        ref %in% c("A", "G") ~ chartr("TCGA", "AGCT", sub),
        TRUE ~ sub),
      trinuc_ref_py = case_when(
        ref %in% c("A", "G") ~ chartr("TCGA", "AGCT", trinuc_ref),
        TRUE ~ trinuc_ref))
}

# function: plot the trinucleotide context of the muts
plot_mut_trinucs <- function(p_dat, p_title = "") {
  sub_colours <-
    c("C>A" = "dodgerblue", "C>G" = "black", "C>T" = "red",
      "T>A" = "grey70", "T>C" = "olivedrab3", "T>G" = "plum2")
  nucs <- c("T", "C", "G", "A")
  py_nucs <- c("C", "T")
  all_subs <-
    expand.grid(do.call(paste0, expand.grid(nucs, py_nucs, nucs)), nucs) %>%
    transmute(trinuc_ref_py = Var1, ref = substr(Var1, 2, 2), alt = Var2,
                     sub_py = paste0(ref, ">", alt)) %>%
    filter(alt != ref)

  p_dat %>%
    mutate(n_cells = n_distinct(id)) %>%
    add_count(mut_id, name = "n_cells_w_mut") %>%
    count(sub_py, trinuc_ref_py) %>%
    right_join(all_subs) %>%
    mutate(n = ifelse(is.na(n), 0, n)) %>%
    ggplot(aes(x = trinuc_ref_py, y = n, fill = sub_py)) +
    geom_col() +
    facet_grid(~ sub_py, scales = "free_x") +
    guides(x = guide_axis(angle = 90)) +
    theme_minimal() +
    theme(axis.text.x = element_text(family = "mono"),
          legend.position = "none") +
    scale_fill_manual(values = sub_colours) +
    scale_y_continuous(expand = c(0, 0)) +
    labs(title = p_title, subtitle = paste(nrow(p_dat), "mutations"),
         x = "", y = "number of mutations")
}

# function: plot VAF heatmap
plot_vaf_heatmap <- function(p_dat, p_title, show_rownames = FALSE) {
  # reshape data for heatmap
  heatmap_data <-
    p_dat %>%
    mutate(mut_id = paste0(ifelse(is.na(gene), "", paste0(gene, "-")), mut_id)) %>%
    add_count(mut_id, name = "n_cells_w_mut") %>%
    reshape2::dcast(mut_id + n_cells_w_mut ~ id, value.var = "alt_vaf")
  rownames(heatmap_data) <- heatmap_data$mut_id
  heatmap_matrix <- as.matrix(heatmap_data[, -c(1:3)]) # drop annotations
  n_total_muts <- n_distinct(p_dat$mut_id)
  n_total_cells <- n_distinct(p_dat$id)

  # calculate the number of mutations per column (id)
  annotation_col <-
    p_dat %>%
    count(id, run_id, celltype, loh_1p, name = "n_muts") %>%
    mutate(run_id = as.character(run_id)) %>%
    tibble::column_to_rownames("id")

  # replace NAs with 0
  heatmap_matrix[is.na(heatmap_matrix)] <- 0

  # prepare annotations for columns
  annotation_row <- data.frame(n = heatmap_data$n)
  rownames(annotation_row) <- heatmap_data$mut_id

  # plot
  pheatmap::pheatmap(
    mat = heatmap_matrix,
    cluster_rows = ifelse(nrow(heatmap_matrix) < 65536, TRUE, FALSE),
    cluster_cols = TRUE,
    show_rownames = show_rownames,
    annotation_row = annotation_row,
    annotation_col = annotation_col,
    color = colorRampPalette(c("white", "blue"))(50),
    main = paste("VAF heatmap:", p_title, "\n",
                 n_total_cells, "cells,", n_total_muts, "mutations"))
}
```

We load mutations generated by previous reports and we combine the calls.

```{r load_vafs}
# load vafs
vafs <-
  c("dna", "dnahyb") %>%
  purrr::set_names() %>%
  purrr::map(~
    system(paste0("ls reports/*_shared_muts_from_bj-somatic-variantcalling_",
           .x, "_filter_cells_cache/vafs_annotated_*.rds"), intern = TRUE) %>%
      readRDS()) %>%
  bind_rows() %>%
  select(-c(id, alt_vaf)) %>%
  group_by(across(-matches("_depth$"))) %>%
  summarise(across(matches("_depth$"), \(x) sum(x, na.rm = TRUE))) %>%
  ungroup() %>%
  mutate(alt_vaf = alt_depth / total_depth) %>%
  # remove common snp vars
  dplyr::filter(!common_snp)
```

We load the samplesheet.

```{r load_data}
man_insp <-
  readr::read_tsv("data/manual_inspection/PD63118.tsv") %>%
  select(-id, -run_id)
metadata <-
  readr::read_csv("data/resolveome/samplesheet_local.csv") %>%
  left_join(man_insp) %>%
  mutate(
    # define celltypes
    celltype = case_when(
      celltype_VDJ_recomb == "alpha-beta T cell" ~ "T cell",
      celltype_VDJ_recomb == "B cell" & celltype_SHM == "mature B cell" ~
        "mature B cell",
      celltype_VDJ_recomb == "B cell" & celltype_SHM == "not mature B cell" ~
        "non-mature B cell",
      TRUE ~ "uncertain"),
    # define loh_1p
    loh_1p = case_when(loh_1p ~ "1p LOH", !loh_1p ~ "no 1p LOH",
                              TRUE ~ "unknown"))
ss <-
  readr::read_csv(file.path(bj_dir, "samplesheet.csv"))
```

We load the driver genes.

```{r driver_genes}
driver_genes <-
  readLines("../trencadis-seq/data/thyroid/driver_genes/driver_genes.txt")
```

# Mutation types per cell

Plot the distribution of mutation types per cell.

```{r plot_mut_types_all, fig.width = 12}
# all muts
vafs %>%
  add_count(id) %>%
  ggplot(aes(x = reorder(id, -n), fill = impact)) +
  geom_bar() +
  scale_fill_manual(values = impact_colours) +
  theme_classic() +
  scale_y_continuous(expand = c(0, 0)) +
  guides(x = guide_axis(angle = -90)) +
  labs(x = "")
```

```{r plot_mut_types_coding, fig.width = 12}
# coding muts
vafs %>%
  filter(impact != "Non-coding") %>%
  add_count(id) %>%
  ggplot(aes(x = reorder(id, -n), fill = impact)) +
  geom_bar() +
  scale_fill_manual(values = impact_colours) +
  theme_classic() +
  scale_y_continuous(expand = c(0, 0)) +
  guides(x = guide_axis(angle = -90)) +
  labs(x = "")
```

# VAF distribution

Plot the VAF distribution.

```{r plot_vaf_dist, fig.height = 4, fig.width = 5}
c(1, 5, 10, 20, 50) %>%
  purrr::walk(function(min_total_depth) {
    p_dat <- vafs %>% filter(total_depth >= min_total_depth)
    p <-
      p_dat %>%
      ggplot(aes(x = alt_vaf)) +
      geom_histogram() +
      theme_classic() +
      labs(title = paste("VAF distribution, min depth =", min_total_depth),
           subtitle = paste(nrow(p_dat), "mutations"), x = "alt VAF")
    print(p)
  })
```

# VAF heatmap

## Plot all mutations (all cells)

```{r plot_vaf_heatmap, fig.height = 6, fig.width = 10}
p <- plot_vaf_heatmap(vafs, "all muts")
print(p)
```

## Plot all shared mutations (all cells)

```{r plot_vaf_heatmap_shared, fig.height = 6, fig.width = 10}
p_dat_shared <-
  vafs %>%
  mutate(n_cells = n_distinct(id)) %>%
  add_count(mut_id, name = "n_cells_w_mut") %>%
  mutate(prop_cells_w_mut = n_cells_w_mut / n_cells) %>%
  # shared mutations
  filter(n_cells_w_mut > 1)
p <- plot_vaf_heatmap(p_dat_shared, "all shared muts")
print(p)
```

## Plot all shared mutations (% cells w mut < 0.3)

In order to exclude some of the germline noise, we look just at mutations 
present in <30% of cells.

```{r plot_vaf_heatmap_shared_filtered, fig.height = 6, fig.width = 10}
p_dat_shared_filtered <-
  p_dat_shared %>%
  filter(prop_cells_w_mut < 0.3)
p <- plot_vaf_heatmap(p_dat_shared_filtered,
                      "shared muts (% cells w mut < 0.3)")
print(p)
```

## Plot shared mutations (common SNPs removed)

```{r plot_vaf_heatmap_shared_no_common, fig.height = 6, fig.width = 10}
p <-
  p_dat_shared %>%
  filter(!common_snp) %>%
  plot_vaf_heatmap("shared muts (common SNPs removed)")
p
```

## Plot coding shared mutations (all cells)

```{r plot_vaf_heatmap_shared_coding, fig.height = 8, fig.width = 15}
# look at coding muts
p_dat_coding <-
  p_dat_shared %>%
  filter(impact != "Non-coding")
p <- plot_vaf_heatmap(p_dat_coding, "shared coding muts (all cells)", TRUE)
print(p)
```

# Mutation spectra

Get the mutational spectra in the cells.

```{r get_spectra}
# get trinucleotide context
vafs_trinuc <-
  xfun::cache_rds({
    vafs %>%
      get_mut_trinucs()
  }, file = "vafs_trinuc.rds", rerun = params$rerun)
```

## Plot trinucleotide spectrum (all cells)

```{r plot_spectra, fig.height = 5, fig.width = 11}
plot_mut_trinucs(vafs_trinuc, "All cells")
```

## Plot trinucleotide spectrum (all cells, shared mutations)

```{r plot_spectra_shared, fig.height = 5, fig.width = 11}
vafs_trinuc %>%
  group_by(chr, pos, ref, alt) %>%
  filter(n() > 1) %>%
  plot_mut_trinucs("All cells, shared mutations")
```

## Plot trinucleotide spectrum per celltype

```{r plot_spectra_per_ct, fig.height = 5, fig.width = 11}
split(vafs_trinuc, vafs_trinuc$celltype) %>%
  purrr::walk(function(vafs_trinuc_i) {
    print(plot_mut_trinucs(vafs_trinuc_i, unique(vafs_trinuc_i$celltype)))
  })
```

## Plot trinucleotide spectrum per cell

```{r plot_spectra_per_cell, fig.height = 5, fig.width = 11}
split(vafs_trinuc, vafs_trinuc$id) %>%
  purrr::walk(function(vafs_trinuc_i) {
    print(plot_mut_trinucs(vafs_trinuc_i, unique(vafs_trinuc_i$id)))
  })
```

# Sequoia

We load the tree.

```{r sequoia}
tree <-
  file.path(bj_run_dir, "/SOMATIC_VARIANT_WORKFLOW_Heuristic_Filter_SEQUOIA/") %>%
  list.files(pattern = "_both_tree_with_branch_length.tree", full.names = TRUE) %>%
  ape::read.tree()
```

## Phylogeny (all cells)

```{r plot_tree, fig.height = 10}
# colour clonal tipes
clones <-
  list(
    clone1 = c("plate3_wellD4_dna_run49882", "plate3_wellE11_dna_run49882"),
    clone2 = c("plate3_wellD11_dna_run49882", "plate3_wellA10_dna_run49882"),
    clone3 = c("plate3_wellD3_dna_run49882", "plate3_wellB7_dna_run49882"),
    clone4 = c("plate3_wellE8_dna_run49882", "plate3_wellF10_dna_run49882"))
tip_col <- rep("black", length(tree$tip.label)) %>% setNames(tree$tip.label)
tip_pal <- RColorBrewer::brewer.pal(length(clones), "Dark2")
purrr::walk2(clones, tip_pal, function(clone, col) {
  tip_col[clone] <<- col
})

# colour celltypes
tip_annotations <-
  metadata %>%
  filter(id %in% tree$tip.label) %>%
  mutate(celltype = ifelse(is.na(celltype), "other", celltype)) %>% 
  {split(.$id, .$celltype)}
tip_annotations <- tip_annotations[tree$tip.label]

# plot tree with colored tip labels
plot(tree, tip.color = tip_col, cex = 0.6)
ape::axisPhylo(side = 1, backward = FALSE)
```

# Driver mutations

```{r drivers}
vafs %>%
  filter(gene %in% driver_genes) %>%
  select(mut_id, gene, id, aachange, ntchange, codonsub, impact,
                ref_depth, alt_depth, alt_vaf) %>%
  knitr::kable()
```

# Case study: plate3_wellD4 + plate3_wellE11

```{r case_study}
clone_vafs <-
  vafs %>%
  group_by(chr, pos, ref, alt) %>%
  mutate(cell_ids = paste(sort(unique(cell_id)), collapse = ",")) %>%
  filter(cell_ids == "plate3_wellD4,plate3_wellE11") %>%
  ungroup()
clone_vafs %>%
  count(chr)
```

# Beta binomial

```{r beta_binom, eval = FALSE}
# get nv and nr
mats <-
  list("NR", "NV") %>%
  purrr::set_names() %>%
  purrr::map(function(mat_type) {
    file.path(bj_run_dir, "SOMATIC_VARIANT_WORKFLOW_Heuristic_Filter_SEQUOIA") %>%
      list.files(pattern = paste0("Mat_", mat_type), full.names = TRUE) %>%
      readr::read_tsv(col_names = FALSE) %>%
      tibble::column_to_rownames("X1") %>%
      as.matrix()
  })

# estimate overdispersion
estimateRho_gridml <- function(NV_vec, NR_vec) {
  # rho will be bounded within 1e-6 and 0.89
 rhovec <- 10^seq(-6, -0.05, by = 0.05)
 mu <- sum(NV_vec) / sum(NR_vec)
 ll <- sapply(rhovec, function(rhoj) sum(dbetabinom(x = NV_vec, size = NR_vec, rho = rhoj, prob = mu, log = TRUE)))
 return(rhovec[ll == max(ll)][1])
}

rho_est <- pvalue <- rep(NA, nrow(mats$NR))
require(VGAM)
for (k in 1:nrow(mats$NR)) {
	rho_est[k] <- estimateRho_gridml(NV_vec = as.numeric(mats$NV[k,]), NR_vec = as.numeric(mats$NR[k,]))
	if (k%%1000 == 0) {
		print(k)
	}
}

# num_samples: those with at least 2 supporting reads
df <-
  data.frame(pos = 1:length(rho_est),
             rho_est = log(rho_est),
             num_samples = rowSums(NV[rownames(NV), ] > 1))

ggplot(df, aes(x = pos, y = rho_est)) +
  geom_point(aes(colour = factor(num_samples), alpha = 0.5)) +
  theme(legend.position = "none")

hist(df$num_samples, breaks = 20, xlab = "Number of samples with mutation",
     ylab = "Number of mutations", main = "pmin(20)")

df$global_vaf <-
  apply(alts[rownames(vafs),], 1, sum) / apply(covs[rownames(vafs), ], 1, sum)
ggplot(df, aes(x = pos, y = rho_est)) +
  geom_point(aes(colour = global_vaf))
```