---
title: "Shared mutations from the nf-resolveome genotyping output"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: true
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
---

```{r setup, include = F}
# rmarkdown::render('reports/20250319_plot_shared_muts_from_nf-resolveome_dna.Rmd', output_file = '20250319_plot_shared_muts_from_nf-resolveome_dna.html', output_dir = 'reports/')

# chunk options
knitr::opts_knit$set(root.dir = "../")
knitr::opts_chunk$set(warning = FALSE,
                      dpi = 600,
                      message = FALSE)

# libraries
library(magrittr)
library(ggplot2)

# function: plot VAF heatmap
plot_vaf_heatmap <- function(p_dat, p_source) {

  # prepare data
  p_dat2 <-
    p_dat %>%
    dplyr::mutate(mut_id = paste(gene, chr, pos, ref, alt, sep = "_")) %>%
    # count number of cells with each mutation
    dplyr::add_count(mut_id, name = "n_cells_w_mut")

  # reshape data for heatmap
  heatmap_data <-
    p_dat2 %>%
    reshape2::dcast(mut_id + n_cells_w_mut ~ id, value.var = "mut_vaf")
  rownames(heatmap_data) <- heatmap_data$mut_id
  heatmap_matrix <- as.matrix(heatmap_data[, -c(1:3)]) # drop annotations
  n_total_muts <- dplyr::n_distinct(p_dat2$mut_id)
  n_total_cells <- dplyr::n_distinct(p_dat2$id)

  # calculate the number of mutations per column (id)
  annotation_col <-
    p_dat2 %>%
    dplyr::count(id, `1p_loh`, celltype_SHM, celltype_VDJ_recomb) %>%
    dplyr::mutate(`1p_loh` = as.character(`1p_loh`)) %>%
    tibble::column_to_rownames("id")

  # replace NAs with 0
  heatmap_matrix[is.na(heatmap_matrix)] <- 0

  # prepare annotations for columns
  annotation_row <- data.frame(n_cells_w_mut = heatmap_data$n_cells_w_mut)
  rownames(annotation_row) <- heatmap_data$mut_id

  # plot
  pheatmap::pheatmap(
    mat = heatmap_matrix,
    cluster_rows = TRUE,
    cluster_cols = TRUE,
    annotation_row = annotation_row,
    annotation_col = annotation_col,
    color = colorRampPalette(c("white", "blue"))(50),
    main = paste(p_source, "- VAF heatmap\n",
                 n_total_cells, "cells,", n_total_muts, "mutations")
  )
}
```

## DNA

We load the samplesheet.

```{r load_data}
man_insp <-
  readr::read_tsv("data/manual_inspection/2024-12-20_PD63118_PTA_BAF_LoH_CellType_Mut_Summary.tsv")
ss <-
  readr::read_csv("out/nf-resolveome/dna/samplesheet.csv") %>%
  dplyr::left_join(man_insp)
```

We load the variants.

```{r load_vars}
geno <-
  readr::read_tsv("out/nf-resolveome/dna/PD63118/genotyping/mutations/PD63118_annotated_mutations.tsv") %>%
  dplyr::left_join(ss)
```

We generate a heatmap of the mutations.

```{r plot_heatmap}
# plot shared mutations
geno %>%
  dplyr::filter(mut_depth > 0) %>%
  dplyr::group_by(chr, pos, ref, alt) %>%
  dplyr::filter(dplyr::n_distinct(id) > 1) %>%
  dplyr::ungroup() %>%
  plot_vaf_heatmap("NanoSeq mutations")
```

## DNA hyb

We load the samplesheet.

```{r load_data}
ss <-
  readr::read_csv("out/nf-resolveome/dnahyb/samplesheet.csv") %>%
  dplyr::left_join(man_insp)
```

We load the variants.

```{r load_vars}
geno <-
  readr::read_tsv("out/nf-resolveome/dnahyb/PD63118/genotyping/mutations/PD63118_annotated_mutations.tsv") %>%
  dplyr::left_join(ss)
```

We generate a heatmap of the mutations.

```{r plot_heatmap}
# plot shared mutations
geno %>%
  dplyr::filter(mut_depth > 0) %>%
  dplyr::group_by(chr, pos, ref, alt) %>%
  dplyr::filter(dplyr::n_distinct(id) > 1) %>%
  dplyr::ungroup() %>%
  plot_vaf_heatmap("NanoSeq mutations")
```