---
title: "Shared mutations from the nf-resolveome genotyping output"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: true
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
---

```{r setup, include = F}
# rmarkdown::render('reports/20250403_plot_shared_muts_from_nf-resolveome_dnahyb.Rmd', output_file = '20250403_plot_shared_muts_from_nf-resolveome_dnahyb.html', output_dir = 'reports/')

# chunk options
knitr::opts_knit$set(root.dir = "../")
knitr::opts_chunk$set(warning = FALSE,
                      dpi = 600,
                      message = FALSE)

# libraries
library(dplyr)
library(magrittr)
library(ggplot2)
source("bin/utils.R")

# dndscv palette
impact_colours <- c("Missense" = "#5f9fa0",
                    "Nonsense" = "#682c8b",
                    "Essential_Splice" = "#8e55a3",
                    "Indels" = "#d4773f",
                    "Synonymous" = "#b3b4b4",
                    "Non-coding" = "#606060")
```

## DNA

We load the samplesheet.

```{r load_data}
data_dir <- "out/nf-resolveome/dnahyb/"
man_insp <-
  readr::read_tsv("data/manual_inspection/PD63118.tsv") %>%
  filter((!chr_dropout | is.na(chr_dropout)) & (!suspected_doublet | is.na(suspected_doublet))) %>%
  mutate(loh_1p = as.character(as.numeric(loh_1p))) %>%
  distinct(cell_id, loh_1p, celltype_SHM, celltype_VDJ_recomb, TNFRSF14_mut)
ss <-
  readr::read_csv(file.path(data_dir, "samplesheet.csv")) %>%
  left_join(man_insp)
```

We load the variants.

```{r load_vars}
genos <-
  readr::read_tsv(file.path(data_dir, "PD63118/genotyping/mutations/PD63118_annotated_mutations.tsv")) %>%
  left_join(ss) %>%
  filter(plate == 3)
```


We generate a heatmap of all mutations.

```{r plot_heatmap, fig.width = 20, fig.height = 20}
p <- plot_vaf_heatmap(genos, show_rownames = FALSE, p_title = "all mutations")
print(p)
```

We generate a heatmap of the shared mutations.

```{r plot_heatmap_shared, fig.width = 20, fig.height = 20}
p <-
  genos %>%
  filter(mut_depth > 0) %>%
  group_by(chr, pos, ref, alt) %>%
  filter(n() > 1) %>%
  plot_vaf_heatmap(show_rownames = FALSE, p_title = "shared mutations")
print(p)
```

We filter to `mut_vaf` > 0.25 and `mut_depth` > 1.

```{r genos_filter}
genos_filter <-
  genos %>%
  filter(mut_vaf > 0.2, mut_depth > 1)

# return
genos_filter
```

```{r plot_heatmap_filtered, fig.width = 20, fig.height = 20}
p <- plot_vaf_heatmap(genos_filter, p_title = "all mutations")
print(p)
```

```{r plot_heatmap_filtered_shared, fig.width = 20, fig.height = 20}
p <- genos_filter %>%
  group_by(chr, pos, ref, alt) %>%
  filter(n() > 1) %>%
  plot_vaf_heatmap(p_title = "shared mutations")
print(p)
```

We generate a heatmap of the nonsynonymous mutations.

```{r plot_nonsym, fig.width = 20, fig.height = 20}
p <-
  genos_filter %>%
  filter(impact != "Synonymous") %>%
  plot_vaf_heatmap("non-synonymous mutations", annotations = "loh_1p")
print(p)
```

We generate a heatmap of the TNFRSF14 mutations.

```{r plot_heatmap_filtered_TNFRSF14, fig.width = 20, fig.height = 20}
p <-
  genos_filter %>%
  filter(gene == "TNFRSF14") %>%
  mutate(alt = paste(alt, aachange)) %>%
  plot_vaf_heatmap("TNFRSF14 mutations")
print(p)
```

We plot the mutations per gene.

```{r muts_per_gene}
p_dat <-
  genos_filter %>%
  filter(impact != "Synonymous") %>%
  group_by(gene) %>%
  summarise(n_cells = n_distinct(cell_id)) %>%
  mutate(prop_cells = n_cells / n_mat_b_cells)
p2 <-
  p_dat %>%
  ggplot(aes(x = reorder(gene, -prop_cells), y = prop_cells)) +
  geom_col(fill = "#d6630d") +
  theme_classic() +
  labs(x = "gene", y = "% of mature B cells with a driver mutation") +
  scale_y_continuous(expand = c(0, 0))

# mutations per gene
p1 <-
  genos_filter %>%
  left_join(man_insp) %>%
  add_count(gene, name = "total") %>%
  count(gene, impact, total, celltype_SHM) %>%
  mutate(impact = ifelse(impact == "no-SNV", "Synonymous", impact) %>% factor(levels = rev(names(impact_colours)))) %>%
  left_join(p_dat %>% distinct(gene, prop_cells)) %>%
  ggplot(aes(x = reorder(gene, -prop_cells), y = n, fill = impact)) +
  geom_col() +
  scale_fill_manual(values = impact_colours) +
  scale_y_continuous(expand = c(0, 0)) +
  theme_classic() +
  theme(axis.text.x = element_blank(), axis.title.x = element_blank(),
        axis.ticks.x = element_blank()) +
  labs(y = "n mutations")

# plot
p1 / p2
```