---
title: "`r params$title`"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: true
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
params:
  title: null
  cache_path: null
---

```{r setup, include = FALSE}
knitr::opts_knit$set(root.dir = "../..")
knitr::opts_chunk$set(
  warning = FALSE,
  message = FALSE,
  dpi = 300,
  cache.path = params$cache_path,
  fig.align = "center"
)
```

```{r libraries, include = F, message = F, warning = F, echo = F}
# libraries
library(magrittr)
library(dplyr)
library(readr)
library(tidyr)
library(ggplot2)
library(tibble)
library(patchwork)
library(GenomicRanges)
library(rtracklayer)
library(liftOver)
library(biomaRt)
library(DESeq2)
library(clusterProfiler)
source("bin/utils.R")

# set ggplot theme
theme_set(theme_classic())

# celltype palette
pal <- c("B" = "#2f9e77",
         "mature B" = "#55c552", "activated B" = "#55c552",
         "non-mature B" = "#397037", "naive B" = "#397037",
         "other" = "#d95f02",
         "NK" = "#e72a8a", "Myeloid" = "#984ea3",
         "non-haematopoietic" = "#ff7f00", "T" = "#377eb8",
         "alpha-beta T cell" = "#377eb8", "B cell" = "#2f9e77",
         "mature B cell" = "#55c552", "not lymphocyte" = "#d95f02",
         "not mature B cell" = "#397037")

# dndscv palette
impact_colours <- c("Missense" = "#5f9fa0",
                    "Nonsense" = "#682c8b",
                    "Essential_Splice" = "#8e55a3",
                    "Indels" = "#d4773f",
                    "no-SNV" = "#d4773f",
                    "Synonymous" = "#b3b4b4",
                    "Non-coding" = "#606060")
```

# Samplesheet

```{r load_ss}
# load manual inspection data
man_insp <-
  read_tsv("data/resolveome/manual_inspection/PD63118.tsv") %>%
  select(-run_id) %>%
  # fix loh_1p
  mutate(loh_1p = as.character(as.numeric(loh_1p)),
         loh_9p = as.character(as.numeric(loh_9p)),
         loh_9q = as.character(as.numeric(loh_9q)))

# load samplesheet
ss <-
  file.path(Sys.getenv("LUSTRE_125"),
            "projects/hashimoto_thyroiditis/data/bams/samplesheet_local.csv") %>%
  read_csv() %>%
  # add manual inspection data
  left_join(man_insp) %>%
  filter(donor_id == "PD63118")

# load list of driver genes
driver_genes <-
  readLines("../trencadis-seq/data/thyroid/driver_genes/driver_genes.txt")
```

# Quality control

We determine which cells to keep. QC has been completed for plate 3, but has not
been completed for plate 10, as only DNAhyb data is available for these cells,
so it is more difficult to definitively call doublets and chromosomal dropouts.
However, based on the new gating strategy for flow cytometry and based on the
pre-PCR quants, we can be fairly confident that the cells are of high quality,
so we will include all plate 10 cells in the analysis.

```{r qc}
# counts
n_seq_cells <- n_distinct(ss$cell_id)
n_seq_dnahyb_cells <- ss %>% filter(seq_type == "dnahyb") %>% {n_distinct(.$cell_id)}
n_seq_dna_cells <- ss %>% filter(seq_type == "dna") %>% {n_distinct(.$cell_id)}

# qc-passing cells
ss_pass <-
  ss %>%
  filter((plate == 3 & !suspected_doublet & !chr_dropout) |
         (plate == 10 &
          (is.na(suspected_doublet) | !suspected_doublet) &
          (is.na(chr_dropout) | !chr_dropout)))
n_pass_cells <- n_distinct(ss_pass$cell_id)
```

Of 263 isolated nuclei, `r n_seq_cells` passed pre-PCR quantification QC
and were submitted for sequencing. `r n_seq_dnahyb_cells` cells underwent 
targeted DNA and RNA sequencing. `r n_seq_dna_cells`/`r n_seq_cells` cells
additionally underwent whole genome DNA sequencing. `r n_pass_cells` cells 
passed QC.

# Celltype annotation

## VDJ + SHM + CSR

```{r get_celltypes, fig.height = 4, fig.width = 6}
cts <-
  ss_pass %>%
  # synthesise celltype annotations from VDJ + SHM + CSR
  mutate(
    celltype_VDJ_SHM_CSR = case_when(
      celltype_VDJ_recomb == "B cell" &
        (celltype_SHM == "mature B cell" |
         class_switch_recombination_CSR == TRUE) ~ "activated B",
      celltype_VDJ_recomb == "B cell" & celltype_SHM == "not mature B cell" ~ "naive B",
      celltype_VDJ_recomb == "alpha-beta T cell" ~ "T",
      celltype_VDJ_recomb == "not lymphocyte" ~ "other",
      TRUE ~ NA) %>%
      factor(levels = c("activated B", "naive B", "T", "other"))) %>%
  filter(!is.na(celltype_VDJ_SHM_CSR)) %>%
  distinct(cell_id, celltype_VDJ_SHM_CSR)

# subset samplesheet
ss_annot <-
  ss_pass %>%
  inner_join(cts)

# plot
cts %>%
  dplyr::count(celltype_VDJ_SHM_CSR) %>%
  ggplot(aes(x = reorder(celltype_VDJ_SHM_CSR, -n), y = n, fill = celltype_VDJ_SHM_CSR)) +
  geom_col() +
  geom_text(aes(label = n), vjust = -0.5) +
  scale_fill_manual(values = pal) +
  theme_classic() +
  theme(legend.position = "none") +
  labs(x = "VDJ + SHM + CSR celltype")

# get stats
ct_stats <-
  cts %>%
  dplyr::count(celltype_VDJ_SHM_CSR) %>%
  {split(.$n, .$celltype_VDJ_SHM_CSR)}
```

Based on VDJ, SHM, and CSR annotations, 
`r ct_stats[["activated B"]] + ct_stats[["naive B"]]` cells were B cells 
(`r ct_stats[["activated B"]]` of which were activated), 
`r ct_stats[["T"]]` were T cells, and `r ct_stats$other` were other celltypes. 

# Sequencing metrics

We load the DNA sequencing coverage.

```{r load_sequencing_metrics}
# get paths to mosdepth output
mosdepth <-
  ss_annot %>%
  transmute(
    seq_type, donor_id, id,
    out_dir = file.path("out/resolveome/nf-resolveome", seq_type, donor_id,
                        "cells", id),
    summary_txt = paste0(out_dir, "/mosdepth/", id, ".mosdepth.summary.txt"),
    dist = paste0(out_dir, "/mosdepth/", id, ".mosdepth.",
                  ifelse(seq_type == "dna", "global", "region"),
                  ".dist.txt")) %>%
  filter(file.exists(summary_txt), file.exists(dist))

# load mean coverage
cov <-
  xfun::cache_rds({
    cov <-
      split(mosdepth$summary_txt, mosdepth$id) %>%
      purrr::map(read_tsv) %>%
      bind_rows(.id = "id") %>%
      dplyr::rename(chr = chrom) %>%
      left_join(ss %>% select(id, seq_type)) %>%
      filter(chr == "total" & seq_type == "dna" |
             chr == "total_region" & seq_type == "dnahyb") %>%
      transmute(id, name = "mean coverage", value = mean)
    cov
  }, file = "cov.rds", rerun = params$rerun)

# load coverage distribution
dist <-
  xfun::cache_rds({
    dist <-
      split(mosdepth$dist, mosdepth$id) %>%
      purrr::map(read_tsv, col_names = c("chr", "cov", "prop")) %>%
      bind_rows(.id = "id") %>%
      filter(chr == "total") %>%
      inner_join(ss_annot)
    dist
  }, file = "dist.rds", rerun = params$rerun)

# get proportion of genome covered per id
perc_cov <-
  dist %>%
  filter(cov == 1) %>%
  transmute(id, name = "% genome covered", value = 100 * prop)
```

## Copy number profiles

```{r plot_ginkgo}

```

## Coverage distribution

```{r plot_coverage_dist}
dist %>%
  filter(prop >= 0.01) %>%
  ggplot(aes(x = cov, y = prop, colour = cell_id)) +
  geom_line() +
  theme_classic() +
  labs(x = "coverage", y = "proportion of genome") +
  facet_grid(~ seq_type, scales = "free_x") +
  theme(legend.position = "none")
```

## Allelic balance of coverage

```{r heterozygosity}
# get genotyped snps
snps <-
  c("dna", "dnahyb") %>%
  purrr::set_names() %>%
  purrr::map(~
    read_tsv(file.path("out/resolveome/nf-resolveome", .x, "PD63118/genotyping",
                       "snps/PD63118_genotyped_snps.tsv"))) %>%
  bind_rows(.id = "seq_type") %>%
  group_by(seq_type) %>%
  # get only runs that pass QC
  filter(id %in% ss_pass$id)

# get binned vafs and depths
p_snps_dat <-
  xfun::cache_rds({
    snps %>%
      mutate(
        mut_vaf = ifelse(total_depth == 0, 0, mut_vaf),
        total_depth_bin = cut(
          total_depth, breaks = c(-Inf, 0, 10, 20, 50, 100, Inf),
          labels = c("0", "<10", "10–20", "20–50", "50–100", ">100")),
        mut_vaf_bin = cut(mut_vaf, breaks = seq(0, 1, length.out = 30 + 1),
                          include.lowest = TRUE),
        mut_vaf_min = stringr::str_extract(mut_vaf_bin, "[\\d\\.]+") %>% as.numeric(),
        mut_vaf_high = as.numeric(stringr::str_extract(mut_vaf_bin, "(?<=,)\\s*[\\d\\.]+")),
        mut_vaf_mid = (mut_vaf_min + mut_vaf_high) / 2) %>%
      dplyr::count(mut_vaf_bin, mut_vaf_mid, total_depth_bin)
  }, file = "p_snps_dat.rds", rerun = params$rerun)

# calculate allelic statuses
allelic_statuses <-
  xfun::cache_rds({
    snps %>%
      mutate(allelic_status = case_when(
        ref_depth > 0 & mut_depth > 0 ~ "both",
        ref_depth > 0 & mut_depth == 0 ~ "one",
        ref_depth == 0 & mut_depth > 0 ~ "one",
        ref_depth == 0 & mut_depth == 0 ~ "neither")) %>%
      dplyr::count(id, allelic_status) %>%
      group_by(seq_type, id) %>%
      mutate(n_snps = sum(n), prop = n / n_snps)
  }, file = "allelic_statuses.rds", rerun = params$rerun)
```

We plot the VAFs of germline heterozygous SNPs across the genome.

```{r plot_allelic_vaf, fig.height = 8}
# get rolling average mut vaf for every 100 snps
snps %>%
  filter(id == "plate10_wellA10_dna_run50382", seq_type == "dna", chr %in% c("22")) %>%
  ggplot(aes(x = pos, y = mut_vaf)) +
  geom_hline(yintercept = 0.5, colour = "grey") +
  ggpointdensity::geom_pointdensity() +
  scale_x_continuous(expand = c(0, 0)) +
  # remove padding between facets
  theme(panel.spacing = unit(0, "lines")) +
  facet_grid(~ chr, scales = "free_x")
dev.off()
```

We plot the VAF distribution of germline heterozygous SNPs.

```{r plot_allelic_vaf_dist, fig.height = 8}
p_snps_dat %>%
  ggplot(aes(x = mut_vaf_mid, y = n, fill = total_depth_bin)) +
  geom_col() +
  scale_fill_viridis_d() +
  theme_classic() +
  labs(title = "VAF distribution", x = "VAF bin", y = "n mutations") +
  lims(x = c(0, 1)) +
  scale_y_continuous(expand = c(0, 0)) +
  facet_grid(seq_type ~ ., scales = "free_y")

p_snps_dat <-
  snps %>%
  filter(id == "plate10_wellA7_dnahyb_run50227") %>%
  mutate(
    mut_vaf = ifelse(total_depth == 0, 0, mut_vaf),
    total_depth_bin = cut(
      total_depth, breaks = c(-Inf, 0, 10, 20, 50, 100, Inf),
      labels = c("0", "<10", "10–20", "20–50", "50–100", ">100")),
    mut_vaf_bin = cut(mut_vaf, breaks = seq(0, 1, length.out = 30 + 1),
                      include.lowest = TRUE),
    mut_vaf_min = stringr::str_extract(mut_vaf_bin, "[\\d\\.]+") %>% as.numeric(),
    mut_vaf_high = as.numeric(stringr::str_extract(mut_vaf_bin, "(?<=,)\\s*[\\d\\.]+")),
    mut_vaf_mid = (mut_vaf_min + mut_vaf_high) / 2) %>%
  dplyr::count(mut_vaf_bin, mut_vaf_mid, total_depth_bin)
p_snps_dat %>%
  ggplot(aes(x = mut_vaf_mid, y = n, fill = total_depth_bin)) +
  geom_col() +
  scale_fill_viridis_d() +
  theme_classic() +
  labs(title = "VAF distribution", x = "VAF bin", y = "n mutations") +
  lims(x = c(0, 1)) +
  scale_y_continuous(expand = c(0, 0)) +
  facet_grid(seq_type ~ ., scales = "free_y")
```

We plot the proportion of allelic dropout per cell.

```{r plot_allelic_dropout, fig.width = 10}
# get prop allelic dropout
allelic_statuses <-
  readRDS("reports/20250709_resolveome_plots_cache/allelic_statuses.rds")
allelic_dropout <-
  allelic_statuses %>%
  filter(allelic_status %in% c("one", "neither")) %>%
  group_by(id, seq_type) %>%
  summarise(name = "allelic dropout", value = sum(prop))

# plot
allelic_statuses %>%
  inner_join(ss_annot %>% select(id, cell_id)) %>%
  left_join(
    allelic_statuses %>%
      filter(seq_type == "dnahyb", allelic_status == "both") %>%
      ungroup() %>%
      select(id, arr = prop)) %>%
  mutate(prop_lab = paste0(round(arr * 100) , "%")) %>%
  ggplot(aes(x = reorder(cell_id, -arr), y = prop, fill = allelic_status)) +
  geom_col(position = "stack") +
  guides(x = guide_axis(angle = -90)) +
  scale_fill_brewer(palette = "Set2") +
  theme_classic() +
  labs(x = "") +
  facet_grid(seq_type ~ .) +
  scale_y_continuous(expand = c(0, 0))
```

We plot the sequencing metrics.

```{r plot_coverage, fig.height = 10, fig.width = 15}
p_cov_dat <-
  bind_rows(cov, perc_cov, allelic_dropout) %>%
  inner_join(ss_annot) %>%
  mutate(name = case_when(seq_type == "dnahyb" & name == "% genome covered" ~ "% panel covered",
                          TRUE ~ name)) %>%
  group_by(cell_id) %>%
  mutate(seq = paste(sort(unique(seq_type)), collapse = " + ")) %>%
  group_by(seq) %>%
  mutate(seq_n = paste0(seq, "\n(n = ", n_distinct(cell_id), ")"))

p1 <-
  p_cov_dat %>%
  left_join(
    bind_rows(
      p_cov_dat %>%
        filter(name == "mean coverage", seq == "dna + dnahyb + rna") %>%
        transmute(cell_id, cov = value) %>%
        distinct(),
      p_cov_dat %>%
        filter(name == "mean coverage", seq == "dnahyb + rna") %>%
        transmute(cell_id, cov = value) %>%
        distinct()
    )) %>%
  ggplot(aes(x = reorder(cell_id, -cov), y = value)) +
  geom_col(position = "dodge") +
  ggh4x::facet_nested(seq_type + name ~ seq_n, scales = "free",
                      space = "free_x", switch = "y") +
  guides(x = guide_axis(angle = 90)) +
  theme_minimal() +
  theme(panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        strip.placement = "outside",
        axis.title.y = element_blank(), axis.title.x = element_blank(),
        strip.background = element_rect(color = "white", fill = "#ededed"))

p2 <-
  p_cov_dat %>%
  ggplot(aes(x = "all", y = value)) +
  geom_boxplot(outlier.colour = NA) +
  geom_jitter(height = 0, width = 0.2, alpha = 0.5, size = 0.5) +
  ggh4x::facet_nested(seq_type + name ~ ., scales = "free",
                      space = "free_x", switch = "y") +
  theme_minimal() +
  scale_y_continuous(position = "right", limits = c(0, NA)) +
  theme(strip.text = element_blank(),
        axis.text.x = element_blank(), axis.title.x = element_blank())

(p1 | p2) + plot_layout(widths = c(10, 1))
```

```{r}
# summarise sequencing stats
seq_stats <-
  p_cov_dat %>%
  group_by(name, seq_type) %>%
  summarise(mean = mean(value), min = min(value), max = max(value)) %>%
  pivot_longer(cols = c(mean, min, max), names_to = "stat") %>%
  split(.$seq_type) %>%
  purrr::map(~ split(.x, .x$name) %>% purrr::map(~ split(.x$value, .x$stat)))

# summarise allelic dropout stats
allelic_statuses_stats <-
  allelic_statuses %>%
  group_by(allelic_status) %>%
  summarise(min = min(prop), max = max(prop), mean = mean(prop)) %>%
  pivot_longer(cols = c(mean, min, max), names_to = "stat") %>%
  mutate(value = round(100 * value, 1)) %>%
  split(.$allelic_status) %>%
  purrr::map(~ split(.x$value, .x$stat)) 
```

Mean proportion coverage was
`r round(seq_stats$dna[["% genome covered"]]$mean)`% (range
`r round(seq_stats$dna[["% genome covered"]]$min)`-
`r round(seq_stats$dna[["% genome covered"]]$max)`%) 
for DNA genome-wide, and
`r round(seq_stats$dnahyb[["% panel covered"]]$mean)`% (range
`r round(seq_stats$dnahyb[["% panel covered"]]$min)`-
`r round(seq_stats$dnahyb[["% panel covered"]]$max)`%) 
for targeted DNA within the immune panel.
Average coverage was
`r round(seq_stats$dna[["mean coverage"]]$mean)`X (range
`r round(seq_stats$dna[["mean coverage"]]$min)`-`r round(seq_stats$dna[["mean coverage"]]$max)`X) 
for the DNA genome-wide, and
`r round(seq_stats$dnahyb[["mean coverage"]]$mean)`X (range
`r round(seq_stats$dnahyb[["mean coverage"]]$min)`-`r round(seq_stats$dnahyb[["mean coverage"]]$max)`X)
for the targeted DNA within the immune panel.

Across heterozygous SNPs, the mean rate of allelic dropout per cell was
`r allelic_statuses_stats$one$mean + allelic_statuses_stats$neither$mean`%. 
This includes sites where only one of the two alleles was read
(mean `r allelic_statuses_stats$one$mean`%,
range `r allelic_statuses_stats$one$min`-`r allelic_statuses_stats$one$max`%)
and sites with no coverage
(mean `r allelic_statuses_stats$neither$mean`%,
range `r allelic_statuses_stats$neither$min`-`r allelic_statuses_stats$neither$max`%).
On average, `r allelic_statuses_stats$both$mean`% of sites had coverage of both
alleles.

# Copy number and LOH profiles

```{r loh_venn}
library(ggvenn)
p_dat <-
  list(
    `1p LOH` = ss_annot %>% filter(loh_1p == "1") %>% pull(cell_id) %>% unique(),
    `9p LOH` = ss_annot %>% filter(loh_9p == "1") %>% pull(cell_id) %>% unique(),
    #`9q LOH` = ss_annot %>% filter(loh_9q == "1") %>% pull(cell_id) %>% unique(),
    `activated B cell` = ss_annot %>% filter(celltype_VDJ_SHM_CSR == "activated B") %>% pull(cell_id) %>% unique())

p_dat %>% ggVennDiagram::ggVennDiagram(label_size = 6)
```

```{r cn_loh}
loh_counts <-
  ss_annot %>%
  distinct(cell_id, loh_1p, loh_9p, loh_9q) %>%
  dplyr::count(loh_1p, loh_9p) %>%
  mutate(loh_lab = paste0("1p=", loh_1p, ",9p=", loh_9p)) %>%
  {split(.$n, .$loh_lab)}

loh_counts_ct_df <-
  ss_annot %>%
  distinct(cell_id, loh_1p, celltype_VDJ_SHM_CSR) %>%
  dplyr::count(loh_1p, celltype_VDJ_SHM_CSR)
loh_counts_ct_df %>% knitr::kable()

loh_counts_ct <-
  loh_counts_ct_df %>%
  split(.$celltype_VDJ_SHM_CSR) %>%
  purrr::map(~ split(.x$n, .x$loh_1p))

n_act_b_loh <- loh_counts_ct[["activated B"]][["1"]]
n_act_b <- sum(unlist(loh_counts_ct[["activated B"]][c("1", "0")]))
perc_act_b_loh <- round(100 * n_act_b_loh / n_act_b)

dna_loh_counts <-
  ss_annot %>%
  filter(seq_type == "dna") %>%
  distinct(cell_id, loh_1p) %>%
  dplyr::count(loh_1p) %>%
  {split(.$n, .$loh_1p)}
```

`r loh_counts[["1p=1,9p=0"]] + loh_counts[["1p=1,9p=1"]]` /
`r sum(unlist(loh_counts))` cells had evidence of 1p LOH. 
`r loh_counts[["1p=0,9p=1"]] + loh_counts[["1p=1,9p=1"]]` /
`r sum(unlist(loh_counts))` cells had evidence of 9p LOH.
`r loh_counts[["1p=1,9p=1"]]` cells had concurrent 1p and 9p LOH. All cells  
in which LOH events were detected were activated B cells [see figure X].
Additionally, 4 cells showed 9q LOH - 2 with 1p LOH and 2 with 1p and 9p LOH.

For cells with whole genome DNA sequencing, there is sufficient resolution in 
the BAF plots to manually place the 1p LOH breakpoints. Of the 
`r dna_loh_counts[["1"]] + dna_loh_counts[["0"]]`
cells with whole genome DNA sequencing,
`r dna_loh_counts[["1"]]` 
cells had evidence of 1p LOH. At least 8 distinct 1p
breakpoints were confirmed among these 
`r dna_loh_counts[["1"]]` 
cells, suggesting 8+ independent 1p copy number events in a population of just 
`r dna_loh_counts[["1"]] + dna_loh_counts[["0"]]`
cells [see figure X].

The Ginkgo copy number calls show no signal of 1p coverage 
loss across these cells [see supplementary figure X], suggesting that these are 
all copy-neutral LOH events. 

# Somatic mutation genotyping

```{r load_vars}
# load the genotyping and pile up reads across DNA and targeted DNA.
genos <-
  ss %>%
  # get genotyping for dna and dnahyb
  filter(seq_type %in% c("dna", "dnahyb")) %>%
  distinct(seq_type, donor_id) %>%
  mutate(genos = paste0("out/resolveome/nf-resolveome/", seq_type, "/", donor_id,
                        "/genotyping/mutations/", donor_id,
                        "_annotated_mutations.tsv")) %>%
  purrr::pmap(function(seq_type, donor_id, genos) {
    read_tsv(genos) %>%
      mutate(donor_id = donor_id) %>%
      # convert id -> cell_id
      inner_join(ss %>% select(id, cell_id)) %>%
      select(-id)
  }) %>%
  # pile up
  bind_rows() %>%
  select(-mut_vaf) %>%
  group_by(across(-matches("_depth$"))) %>%
  summarise(across(matches("_depth$"), \(x) sum(x, na.rm = TRUE))) %>%
  ungroup() %>%
  mutate(mut_vaf = mut_depth / total_depth) %>%
  # annotate mutations on 1p
  mutate(on_1p = ifelse(chr == "1" & pos < 123400000, TRUE, FALSE)) %>%
  # filter to only annotated, qc'ed cells
  inner_join(ss_annot %>% distinct(cell_id, celltype_VDJ_SHM_CSR, loh_1p, loh_9p)) %>%
  # fix no-snv annotation
  mutate(impact = ifelse(impact == "no-SNV" & type %in% c("ins", "del"), "Indels", impact))

# return
genos
```

We plot the distribution of genotyped mutations from NanoSeq across the genes.

```{r plot_genos, fig.height = 2, fig.width = 9}
load("../reference/dndscv/refcds_hg19.rda")

# plot CDS mutations in drivers
driver_genes %>%
  purrr::map(function(gene_i) {

    # get ccds
    g <- RefCDS[[which(purrr::map_lgl(RefCDS, ~ .x$gene_name == gene_i))]]
    ccds <-
      tibble::tibble(chr = g$chr,
                     start = g$intervals_cds[, 1],
                     end = g$intervals_cds[, 2]) %>%
      mutate(cds_interval = row_number()) %>%
      group_by(chr, start, end, cds_interval) %>%
      reframe(pos = start:end) %>%
      arrange(chr, pos) %>%
      mutate(aa = (row_number() - 1) %/% 3 + 1,
             cds_boundary = cds_interval != lag(cds_interval) | row_number() == 1,
             cds_end = row_number() == n(),
             aa_boundary = case_when(cds_boundary ~ aa - 0.5,
                                     cds_end ~ aa + 0.5,
                                     TRUE ~ NA))

    # plot
    p_muts <-
      genos %>%
      filter(gene == gene_i) %>%
      distinct(chr, pos, ref, alt, gene, strand, ref_cod, mut_cod, ref3_cod,
               mut3_cod, impact, ntchange, codonsub, pid, type) %>%
      mutate(impact = ifelse(impact == "no-SNV" & type %in% c("ins", "del"),
                             "Indels", impact))
    p_dat <-
      p_muts %>%
      right_join(ccds) %>%
      dplyr::count(chr, pos, ref, alt, impact, aa, aa_boundary)

    # stats
    n_muts <- format(nrow(p_muts), big.mark = ",", scientific = FALSE)
    n_cds_muts <-
      p_dat %>%
      filter(!is.na(impact)) %>%
      nrow() %>%
      format(big.mark = ",", scientific = FALSE)

    # plot
    p_dat %>%
      ggplot(aes(x = aa, y = n, fill = impact)) +
      geom_vline(aes(xintercept = aa_boundary), colour = "grey") +
      geom_col(data = p_dat %>% filter(!is.na(impact)), width = 1,
               position = "stack") +
      scale_fill_manual(values = impact_colours) +
      theme_classic() +
      theme(panel.grid.major.x = element_blank(),
            panel.grid.minor.x = element_blank(),
            axis.line.y = element_blank()) +
      scale_y_continuous(expand = c(0, 0)) +
      scale_x_continuous(expand = c(0, 0)) +
      labs(title = paste0(gene_i, " CDS mutations (1 aa bins, n = ", n_muts,
                          ", n in CDS = ", n_cds_muts, ")"))

  })
```

We filter to `mut_vaf` > 0.2 and `mut_depth` > 1.

```{r genos_filter}
# filter
genos_filter <-
  genos %>%
  filter(mut_vaf > 0.2, mut_depth > 1)

# save
ss_annot %>%
  select(cell_id, celltype_VDJ_SHM_CSR) %>%
  left_join(
    genos_filter %>%
      filter(gene == "TNFRSF14", impact != "Synonymous") %>%
      transmute(cell_id, TNFRSF14_mut = TRUE) %>%
      distinct()) %>%
  mutate(TNFRSF14_mut = ifelse(is.na(TNFRSF14_mut), FALSE, TRUE)) %>%
  distinct(cell_id, celltype_VDJ_SHM_CSR, TNFRSF14_mut) %>%
  readr::write_tsv("out/analysis/PD63118_TNFRSF14_mut_status.tsv")
genos_filter %>%
  readr::write_tsv("out/analysis/PD63118_filtered_genotyped_mutations.tsv")

# return
genos_filter
```

We plot the co-occurrence of CD274 + 9p LOH and TNFRSF14 + 1p LOH.

```{r plot_cooccur}
cooccur <-
  genos_filter %>%
  group_by(cell_id) %>%
  summarise(`TNFRSF14 mut` = any(gene == "TNFRSF14"),
            `CD274 mut` = any(gene == "CD274")) %>%
  left_join(
    ss_annot %>%
      transmute(cell_id,
                       `1p LOH` = as.logical(as.numeric(loh_1p)),
                       `9p LOH` = as.logical(as.numeric(loh_9p))) %>%
      distinct()) %>%
  pivot_longer(cols = -c("cell_id")) %>%
  filter(value) %>%
  {split(.$cell_id, .$name)}

ggVennDiagram::ggVennDiagram(cooccur)
```

```{r genotyping_stats, fig.width = 15, fig.height = 7}
# how many mutant sites across cells had reads?
prop_muts_covered <-
  genos %>%
  mutate(has_cov = total_depth > 0, cov_over_5 = total_depth >= 5) %>%
  group_by(cell_id) %>%
  summarise(perc_covered = 100 * sum(has_cov) / n()) %>%
  summarise(mean = mean(perc_covered), min = min(perc_covered),
            max = max(perc_covered)) %>%
  as.list()

# plot by depth
p_dat <-
  genos %>%
  mutate(
    `total depth bin` = cut(
      total_depth, breaks = c(-Inf, 0, 10, 20, 50, 100, Inf),
      labels = c("0", "<10", "10–20", "20–50", "50–100", ">100"))) %>%
  dplyr::count(cell_id, `total depth bin`, name = "n genotyped sites")
p_dat %>%
  left_join(
    p_dat %>%
      filter(`total depth bin` == ">100") %>%
      transmute(cell_id, n_100 = `n genotyped sites`)) %>%
  ggplot(aes(x = reorder(cell_id, -n_100), y = `n genotyped sites`,
             colour = `total depth bin`, fill = `total depth bin`)) +
  geom_col(position = "stack") +
  viridis::scale_fill_viridis(discrete = TRUE) +
  viridis::scale_colour_viridis(discrete = TRUE) +
  theme(legend.position = "none") +
  theme_classic() +
  guides(x = guide_axis(angle = -90)) +
  scale_y_continuous(expand = c(0, 0)) +
  theme(axis.title.x = element_blank())

# how many mutations were genotyped?
n_muts <- genos %>% distinct(chr, pos, ref, alt) %>% nrow() %>% format(big.mark = ",", scientific = FALSE)
n_muts_geno <- genos_filter %>% distinct(chr, pos, ref, alt) %>% nrow() %>% format(big.mark = ",", scientific = FALSE)
n_genes <- n_distinct(genos$gene) %>% format(big.mark = ",", scientific = FALSE)
n_genes_geno <- n_distinct(genos_filter$gene) %>% format(big.mark = ",", scientific = FALSE)

# which celltypes are mutated?
genos_filter %>% dplyr::count(celltype_VDJ_SHM_CSR)

# what are the top most mutated genes?
top_mut_genes <-
  genos_filter %>% dplyr::count(celltype_VDJ_SHM_CSR, gene) %>% arrange(-n)
top_mut_genes %>% knitr::kable()

# how many cells and what types of mutations do they have?
top_mut_genes_stats <-
  genos_filter %>%
  group_by(gene, impact) %>%
  mutate(n_mut_per_impact = n(),
         n_mut_per_impact_lab = paste(n(), impact)) %>%
  group_by(gene) %>%
  arrange(-n_mut_per_impact) %>%
  summarise(n_mut_cells = n_distinct(cell_id),
            n_mut = n(),
            impact = paste(unique(n_mut_per_impact_lab), collapse = ", ")) %>%
  arrange(-n_mut) %>%
  mutate(stat = paste0(gene, " (n = ", n_mut,
                       ", n mutant cells = ", n_mut_cells, "; ",
                       impact, ")"))
```

DNA and targeted DNA genotyping counts were piled up per cell.

The mean percentage of genotyped sites covered by at least 1 read was 
`r round(prop_muts_covered$mean, 1)`%
(range `r round(prop_muts_covered$min, 1)`-`r round(prop_muts_covered$max, 1)`%) 
[see figure X].

`r n_muts_geno` / `r n_muts` NanoSeq mutations in `r n_genes_geno` / `r n_genes`
genes were successfully genotyped in this population of cells (total n mutations 
across cells = `r nrow(genos_filter)`).
All `r format(n_muts_geno, big.mark = ",", scientific = FALSE)` mutations were
found exclusively in activated B cells [see figure X]. 

The top mutated genes were 
`r top_mut_genes_stats[1, ]$stat`,
`r top_mut_genes_stats[2, ]$stat`, and
`r top_mut_genes_stats[3, ]$stat`.

```{r plot_vaf_dist_hyb, fig.width = 8, fig.height = 7}
# % of B cells with a nonsynonymous mutation
n_activ_b_cells <-
  ss_annot %>%
  filter(celltype_VDJ_SHM_CSR == "activated B") %>%
  pull(cell_id) %>%
  n_distinct()

p_dat <-
  list(
    `% of activated B cells with a driver` =
      genos_filter %>%
      filter(impact != "Synonymous") %>%
      group_by(gene) %>%
      summarise(n_cells = n_distinct(cell_id)) %>%
      transmute(gene, n = n_cells / n_activ_b_cells, impact = "n"),
    `n NanoSeq mutations per gene` =
      genos %>%
      distinct(chr, pos, ref, alt, impact, gene) %>%
      dplyr::count(gene, impact),
    `n ResolveOME mutations per gene` =
      genos_filter %>%
      dplyr::count(gene, impact))

p_dat %>%
  bind_rows(.id = "name") %>%
  mutate(name = factor(name, levels = c("n NanoSeq mutations per gene", "n ResolveOME mutations per gene", "% of activated B cells with a driver")),
         impact = factor(impact, levels = c(names(impact_colours), "n"))) %>%
  right_join(
    p_dat$`% of activated B cells with a driver` %>% select(gene, prop = n)) %>%
  ggplot(aes(x = reorder(gene, -prop), y = n, fill = impact)) +
  geom_col(position = "stack") +
  theme_minimal() +
  theme(strip.placement = "outside", axis.title.y = element_blank(),
        axis.title.x = element_blank(), panel.grid.major.x = element_blank()) +
  scale_y_continuous(expand = c(0, 0)) +
  scale_fill_manual(values = c(impact_colours, c("n" = "#d6630d"))) +
  facet_grid(name ~ ., scales = "free_y", switch = "y") +
  guides(x = guide_axis(angle = -45))
```

We plot the impact of the mutations genotyped.

```{r plot_impact}
genos_filter %>%
  distinct(chr, pos, ref, alt, impact, gene) %>%
  add_count(gene) %>%
  ggplot(aes(x = reorder(gene, -n), fill = impact)) +
  geom_bar() +
  geom_text(aes(label = n, y = n), vjust = -0.5) +
  scale_fill_manual(values = impact_colours) +
  guides(x = guide_axis(angle = -90)) +
  theme_classic() +
  labs(x = "", y = "n mutations")
```

We generate a heatmap of all mutations.

```{r plot_heatmap, fig.width = 20, fig.height = 20}
plot_vaf_heatmap(genos, p_title = "all mutations", show_rownames = FALSE,
                 annotations = c("celltype_VDJ_SHM_CSR", "loh_1p"))
```

```{r plot_heatmap_filtered, fig.width = 20, fig.height = 20}
plot_vaf_heatmap(genos_filter, p_title = "filtered mutations",
                 show_rownames = FALSE,
                 annotations = c("celltype_VDJ_SHM_CSR", "loh_1p"))
```

```{r plot_heatmap_filtered_shared, fig.width = 20, fig.height = 20}
p <-
  genos_filter %>%
  group_by(chr, pos, ref, alt) %>%
  filter(n() > 1) %>%
  plot_vaf_heatmap(p_title = "filtered shared mutations",
                   annotations = c("celltype_VDJ_SHM_CSR", "loh_1p"))
print(p)
```

```{r plot_heatmap_filtered_nonsynonymous, fig.width = 20, fig.height = 20}
p <-
  genos_filter %>%
  filter(impact != "Synonymous") %>%
  plot_vaf_heatmap("filtered non-synonymous mutations",
                   annotations = c("celltype_VDJ_SHM_CSR", "loh_1p"))
print(p)
```

```{r plot_heatmap_filtered_TNFRSF14_nonsynonymous, fig.width = 15, fig.height = 10}
p <-
  genos_filter %>%
  filter(gene == "TNFRSF14", impact != "Synonymous") %>%
  plot_vaf_heatmap("TNFRSF14 mutations",
                   annotations = c("celltype_VDJ_SHM_CSR", "loh_1p"))
print(p)
```

```{r plot_heatmap_TNFRSF14_CD274, fig.width = 13, fig.height = 10}
p <-
  genos_filter %>%
  filter(gene %in% c("TNFRSF14", "CD274"), impact != "Synonymous") %>%
  mutate(alt = paste(alt, aachange)) %>%
  plot_vaf_heatmap("non-synonymous TNFRSF14 and CD274 mutations",
                   annotations = c("celltype_VDJ_SHM_CSR", "loh_1p", "loh_9p"))
print(p)
```

We plot TNFRSF14 mutant calls in the 1p LOH and non-1p LOH cells.

```{r}
p <-
  genos_filter %>%
  filter(gene == "TNFRSF14") %>%
  ggplot(aes(x = loh_1p, y = mut_vaf)) +
  geom_boxplot(outlier.colour = NA) +
  geom_jitter(height = 0, width = 0.1) +
  theme_classic()
print(p)
```

We plot the VAF distribution of the mutations.

```{r}
genos %>%
  left_join(ss) %>%
  filter(mut_vaf > 0) %>%
  ggplot(aes(x = mut_vaf, fill = on_1p & loh_1p == "1")) +
  geom_histogram() +
  theme_classic() +
  scale_y_continuous(expand = c(0, 0)) +
  scale_fill_brewer(palette = "Set1")
genos_filter %>%
  left_join(ss) %>%
  ggplot(aes(x = mut_vaf, fill = on_1p & loh_1p == "1")) +
  geom_histogram() +
  lims(x = c(0, NA)) +
  theme_classic() +
  scale_y_continuous(expand = c(0, 0)) +
  scale_fill_brewer(palette = "Set1")
genos_filter %>%
  mutate(groups = case_when(on_1p & loh_1p == "1" ~ "1p", TRUE ~ "other")) %>%
  group_by(groups) %>%
  summarise(mean = mean(mut_vaf, na.rm = T), min = min(mut_vaf, na.rm = T),
            max = max(mut_vaf, na.rm = T))
```

```{r plot_drivers_per_cell}
p_dat <-
  genos_filter %>%
  filter(impact != "Synonymous") %>%
  arrange(loh_1p) %>%
  mutate(cell_id = forcats::fct_inorder(cell_id)) %>%
  add_count(gene) %>%
  add_count(gene, cell_id, name = "n per cell") %>%
  mutate(gene = paste0(gene, " (", n, ")")) %>%
  arrange(n) %>%
  mutate(gene = forcats::fct_inorder(gene)) %>%
  distinct(cell_id, celltype_VDJ_SHM_CSR, gene, loh_1p, `n per cell`)

p1 <-
  p_dat %>%
  ggplot(aes(x = cell_id, y = 1, fill = loh_1p)) +
  geom_tile() +
  scale_fill_brewer(palette = "Set1") +
  theme_void()

p2 <-
  p_dat %>%
  ggplot(aes(x = cell_id, y = gene, fill = `n per cell`)) +
  geom_tile() +
  theme_classic() +
  guides(x = guide_axis(angle = -90)) +
  viridis::scale_fill_viridis() +
  coord_equal()

p1 / p2 + plot_layout(heights = c(1, 10), guides = "collect")
```

# Somatic mutation calling

We load mutation calls from `bj-somatic-variantcalling` and we combine the calls
per cell.

```{r load_vafs}
bj_muts_per_seq_type <- xfun::cache_rds({

  # get NR and NV
  bj_muts_per_seq_type_38 <-
    c("dna", "dnahyb") %>%
    purrr::set_names() %>%
    purrr::map(function(seq_type) {
      # get nr and nv
      c("mut_depth" = "NV", "total_depth" = "NR") %>%
        purrr::map(function(i) {
          paste0("out/resolveome/basejumper/bj-somatic-variantcalling/", seq_type,
                "/PD63118_run/SOMATIC_VARIANT_WORKFLOW_Heuristic_Filter_SEQUOIA", 
                "/Sequoia_group_PD63118_bino-10_rhosnp0.4_rhoindel0.4_mincov10_maxcov500_both_",
                i, "_filtered_all.txt") %>%
            read.table() %>%
            tibble::rownames_to_column(var = "mut_id")
        }) %>%
        bind_rows(.id = "name") %>%
        pivot_longer(cols = -c("name", "mut_id"), names_to = "id") %>%
        pivot_wider() %>%
        separate_wider_delim(cols = "mut_id", delim = "_", cols_remove = FALSE,
                            names = c("chr", "pos", "ref", "alt")) %>%
        mutate(pos = as.numeric(pos)) %>%
        # calculate global VAFs
        group_by(mut_id) %>%
        mutate(global_mut_depth = sum(mut_depth),
               global_total_depth = sum(total_depth),
               global_mut_vaf = global_mut_depth / global_total_depth) %>%
        ungroup() %>%
        # get muts only
        filter(mut_depth > 0)
    }) %>%
    bind_rows(.id = "seq_type") %>%
    # filter to only annotated, qc'ed cells
    inner_join(ss_annot %>% distinct(id, cell_id, celltype_VDJ_SHM_CSR, loh_1p))

  # lift over to grch37
  bj_muts_per_seq_type <-
    bj_muts_per_seq_type_38 %>%
    {GRanges(seqnames = .$chr,
             ranges = IRanges(start = .$pos, end = .$pos),
             pos = .$pos)} %>%
    liftOver(import.chain("../reference/liftOver/hg38ToHg19.over.chain")) %>%
    unlist() %>%
    tibble::as_tibble() %>%
    distinct() %>%
    transmute(chr = seqnames, pos_37 = start, pos) %>%
    inner_join(bj_muts_per_seq_type_38) %>%
    mutate(pos = pos_37, mut_id = paste(chr, pos, ref, alt, sep = "_")) %>%
    select(-pos_37)

  # annotate bj_muts with mutation impacts using dndscv
  bj_muts_per_seq_type <-
    bj_muts_per_seq_type %>%
    left_join(
      bj_muts_per_seq_type %>%
        transmute(sampleID = "", chr = gsub("chr", "", chr), pos, ref, alt) %>%
        distinct() %>%
        dndscv::dndscv(max_muts_per_gene_per_sample = Inf, outp = 1,
                       max_coding_muts_per_sample = Inf) %>%
        {.$annotmuts} %>%
        mutate(chr = paste0("chr", chr)) %>%
        select(-sampleID) %>%
        # collapse annotations on overlapping genes
        group_by(across(-c(gene, pid))) %>%
        summarise(gene = paste(sort(unique(gene)), collapse = ", "),
                  pid = paste(sort(unique(pid)), collapse = ", ")) %>%
        ungroup())

  #return
  bj_muts_per_seq_type

}, file = "bj_muts_per_seq_type.rds", rerun = params$rerun)
```

We visualise the global VAF distribution.

```{r plot_global_vaf}
bj_muts_per_seq_type %>%
  distinct(chr, mut_id, seq_type, global_mut_vaf) %>%
  ggplot(aes(x = global_mut_vaf)) +
  geom_histogram(bins = 100) +
  lims(x = c(NA, 1)) +
  ggh4x::facet_nested(seq_type ~ ., scales = "free_y") +
  scale_y_continuous(trans = scales::pseudo_log_trans()) +
  theme_bw()
```

We filter out mutations with a global mutant VAF > 0.25.

```{r global_vaf_filter}
bj_muts_per_seq_type_filter <-
  bj_muts_per_seq_type %>%
  filter(global_mut_vaf < 0.25)
```

We visualise the global VAF distribution after filtering.

```{r plot_global_vaf_filter}
bj_muts_per_seq_type_filter %>%
  distinct(chr, mut_id, seq_type, global_mut_vaf) %>%
  ggplot(aes(x = global_mut_vaf)) +
  geom_histogram(bins = 100) +
  lims(x = c(NA, 1)) +
  ggh4x::facet_nested(seq_type ~ ., scales = "free_y") +
  scale_y_continuous(trans = scales::pseudo_log_trans()) +
  theme_bw()
```

```{r get_bj_muts}
  # pileup across dna and dnahyb
bj_muts <- xfun::cache_rds({
  bj_muts_per_seq_type_filter %>%
    group_by(cell_id) %>%
    mutate(seq_type = paste(sort(unique(seq_type)), collapse = "+")) %>%
    group_by(across(-c(id, ends_with("_depth"), global_mut_vaf))) %>%
    #group_by(chr, pos, ref, alt, mut_id, cell_id, seq_type) %>%
    summarise(mut_depth = sum(mut_depth), total_depth = sum(total_depth)) %>%
    ungroup() %>%
    mutate(mut_vaf = mut_depth / total_depth)
}, file = "bj_muts.rds", rerun = params$rerun)
```

We filter to `mut_vaf` > 0.3 and `mut_depth` > 5, and present in less than 50% 
of cells.

```{r filter_calls}
bj_muts_filter <-
  xfun::cache_rds({
    bj_muts %>%
      filter(mut_vaf > 0.3, mut_depth > 5) %>%
      group_by(chr, pos, ref, alt) %>%
      mutate(n_cells_w_mut = n_distinct(cell_id),
            n_cells = n_distinct(bj_muts$cell_id)) %>%
      filter(n_cells_w_mut / n_cells < 0.5) %>%
      ungroup()
  }, file = "bj_muts_filter.rds", rerun = params$rerun)
```

Plot the number of mutations per cell.

```{r plot_muts_per_cell, fig.width = 10, fig.height = 6}
bj_muts %>%
  left_join(bj_muts_filter %>% distinct(mut_id, cell_id) %>% mutate(filter = "pass")) %>%
  mutate(filter = ifelse(is.na(filter), "fail", filter)) %>%
  add_count(cell_id, name = "total") %>%
  dplyr::count(cell_id, filter, , seq_type, total) %>%
  ggplot(aes(x = reorder(cell_id, -total), y = n, fill = filter)) +
  geom_col(position = "stack") +
  ggh4x::facet_grid2(~ seq_type, scales = "free", space = "free_x", independent = "y") +
  guides(x = guide_axis(angle = -90)) +
  theme_classic() +
  scale_y_continuous(expand = c(0, 0)) +
  scale_fill_brewer(palette = "Set1")
```

Plot the distribution of mutation types per cell.

```{r plot_mut_types_all, fig.width = 14, fig.height = 10}
p_dat <-
  bind_rows(
    bj_muts_filter %>%
      mutate(group = "all mutations",
            impact = ifelse(is.na(impact), "Non-coding", impact)),
    bj_muts_filter %>%
      filter(impact != "Non-coding") %>%
      mutate(group = "coding mutations"),
    bj_muts_filter %>%
      transmute(cell_id, celltype = celltype_VDJ_SHM_CSR, seq_type,
                group = "celltype") %>%
      distinct()) %>%
  mutate(impact = factor(impact, levels = names(impact_colours)),
         group = factor(group, levels = c("celltype", "all mutations", "coding mutations"))) %>%
  left_join(bj_muts_filter %>% dplyr::count(cell_id, seq_type))

p_dat %>%
  ggplot(aes(x = reorder(cell_id, -n))) +
  geom_tile(aes(y = 0.5, fill = celltype)) +
  geom_bar(aes(fill = impact)) +
  scale_fill_manual(values = c(impact_colours, pal), na.value = NA,
                    na.translate = FALSE) +
  ggh4x::facet_grid2(group ~ seq_type, scales = "free", space = "free_x",
                     independent = "y") +
  theme_classic() +
  scale_y_continuous(expand = c(0, 0)) +
  guides(x = guide_axis(angle = -90)) +
  labs(x = "", y = "n mutations") +
  ggh4x::facetted_pos_scales(
    y = list(
      group %in% c("all mutations", "coding mutations") ~
        scale_y_continuous(expand = c(0, 0)),
      group == "celltype" ~ scale_y_continuous(labels = NULL, breaks = NULL)))
```

Plot the VAF distribution.

```{r plot_vaf_dist, fig.height = 6, fig.width = 9}
bj_muts %>%
  plot_vaf_dist()
bj_muts_filter %>%
  plot_vaf_dist()
```

Plot the VAF heatmap.

```{r plot_vaf_heatmap_bj_muts, fig.width = 10, fig.height = 8}
bj_muts_filter %>%
  filter(impact != "Non-coding") %>%
  plot_vaf_heatmap(show_rownames = FALSE, annotations = c("loh_1p", "celltype_VDJ_SHM_CSR"))
```

```{r bj_muts_stats}
# n mutated cells
n_mut_cells <- n_distinct(bj_muts_filter$cell_id)
n_cells <- n_distinct(ss_annot$cell_id)

# n mutations
n_muts <- nrow(bj_muts_filter) %>% format(big.mark = ",", scientific = FALSE)

# n coding mutations
n_coding_muts <-
  bj_muts_filter %>%
  filter(impact != "Non-coding") %>%
  nrow() %>%
  format(big.mark = ",", scientific = FALSE)

# n private mutations
n_private_muts <-
  bj_muts_filter %>%
  group_by(chr, pos, ref, alt) %>%
  filter(n_distinct(cell_id) == 1) %>%
  nrow() %>%
  format(big.mark = ",", scientific = FALSE)

# n private coding mutations
n_private_coding_muts <-
  bj_muts_filter %>%
  filter(impact != "Non-coding") %>%
  group_by(chr, pos, ref, alt) %>%
  filter(n_distinct(cell_id) == 1) %>%
  nrow() %>%
  format(big.mark = ",", scientific = FALSE)

# n shared mutations
n_shared_muts <-
  bj_muts_filter %>%
  group_by(chr, pos, ref, alt) %>%
  filter(n_distinct(cell_id) > 1) %>%
  distinct(chr, pos, ref, alt) %>%
  nrow() %>%
  format(big.mark = ",", scientific = FALSE)

# n shared coding mutations
n_shared_coding_muts <-
  bj_muts_filter %>%
  filter(impact != "Non-coding") %>%
  group_by(chr, pos, ref, alt) %>%
  filter(n_distinct(cell_id) > 1) %>%
  distinct(chr, pos, ref, alt) %>%
  nrow() %>%
  format(big.mark = ",", scientific = FALSE)

# mean muts per cell
mean_muts_per_cell <-
  bj_muts_filter %>%
  filter(seq_type == "dna+dnahyb") %>%
  dplyr::count(cell_id) %>%
  summarise(mean = mean(n), min = min(n), max = max(n)) %>%
  as.list() %>%
  purrr::map(~ round(.x) %>% format(big.mark = ",", scientific = FALSE))

# how many of the genotyped mutations were called?
muts_genos_overlap <-
  full_join(
    bj_muts_filter %>%
      transmute(chr = gsub("^chr", "", chr), pos, ref, alt, cell_id,
                bj_muts = "bj_muts") %>%
      distinct(),
    genos_filter %>%
      transmute(chr, pos, ref, alt, cell_id, gene, genos = "genos") %>%
      distinct()) %>%
  dplyr::count(bj_muts, genos) %>%
  mutate(overlap = paste(bj_muts, genos)) %>%
  {split(.$n, .$overlap)}
```

DNA and targeted DNA calls from `bj-somatic-variantcalling` were piled up per
cell. `r n_muts` mutations were called across `r n_mut_cells` / `r n_cells` 
cells, of  which `r n_coding_muts` were coding. Most (`r n_private_muts`) of 
these were private to just one cell, while `r n_shared_muts` mutations were 
shared by at least two cells.

Among cells with whole genome sequencing, the mean burden per cell was
`r mean_muts_per_cell$mean` (range `r mean_muts_per_cell$min`-`r mean_muts_per_cell$max`).

To check the sensitivity of this pipeline, I checked how many of the 
high-confidence genotyped NanoSeq mutations were also called here.
`r muts_genos_overlap[["bj_muts genos"]]` / 
`r muts_genos_overlap[["NA genos"]] + muts_genos_overlap[["bj_muts genos"]]`
high-confidence genotyped mutations were independently called by this pipeline,
indicating a certain degree of false negatives in the call set.

# Mutational spectra

Get the trinucleotide context of each mutation.

```{r get_spectra}
# get trinucleotide context
bj_muts_trinuc <-
  xfun::cache_rds({
    bj_muts_trinuc <-
      bj_muts_filter %>%
      get_mut_trinucs()
    bj_muts_trinuc
  }, file = "bj_muts_trinuc.rds", rerun = params$rerun)
```

Plot trinucleotide spectrum.

```{r plot_spectra_good, fig.height = 5, fig.width = 11}
bj_muts %>%
  get_mut_trinucs() %>%
  plot_mut_trinucs("all bj-somatic-variantcalling calls")
bj_muts_filter %>%
  get_mut_trinucs() %>%
  plot_mut_trinucs("filtered bj-somatic-variantcalling calls")
bj_muts %>%
  group_by(mut_id) %>%
  filter(n_distinct(cell_id) > 1) %>%
  ungroup() %>%
  get_mut_trinucs() %>%
  plot_mut_trinucs("shared bj-somatic-variantcalling calls")
```

# Phylogeny

```{r phylo}
tree <-
  c("dna", "dnahyb") %>%
  purrr::set_names() %>%
  purrr::map(function(seq_type) {
    # read tree
    tree <-
      file.path("out/resolveome/basejumper/bj-somatic-variantcalling/", seq_type,
              "/PD63118_run/SOMATIC_VARIANT_WORKFLOW_Heuristic_Filter_SEQUOIA/") %>%
      list.files(pattern = "_both_tree_with_branch_length.tree", full.names = TRUE) %>%
      ape::read.tree()

    # subset cells
    ape::drop.tip(tree, setdiff(tree$tip.label, unique(ss_annot$id)))
  })

# get tip labels
plot_tree <- function(p_tree) {

  get_tip_annot <- function(...) {
    tips_annot %>%
      filter(...) %>%
      pull(n)
  }

  # get tip annotations
  tips_annot <-
    tibble(id = p_tree$tip.label) %>%
    left_join(ss_annot) %>%
    mutate(n = row_number()) %>%
    # tnfrsf14 mutations
    left_join(
      genos_filter %>%
        filter(gene == "TNFRSF14", !impact %in% c("Synonymous", "Non-coding")) %>%
        transmute(cell_id, tnfrsf14_mut = TRUE) %>%
        distinct()) %>%
    # cd274 mutations
    left_join(
      genos_filter %>%
        filter(gene == "CD274", !impact %in% c("Synonymous", "Non-coding")) %>%
        transmute(cell_id, cd274_mut = TRUE) %>%
        # add in CD274 exon 5 & 6 deletion in plate3_wellD4 (observed from IGB)
        bind_rows(tibble(cell_id = "plate3_wellD4", cd274_mut = TRUE)) %>%
        distinct())
  
  # fix tip labels
  p_tree$tip.label <- gsub("_dna.*", "", p_tree$tip.label)

  # get tip annotations
  tips <- list()
  tips$loh_1p <- get_tip_annot(loh_1p == "1")
  tips$activ_b <- get_tip_annot(celltype_VDJ_SHM_CSR == "activated B")
  tips$tnfrsf14 <- get_tip_annot(tnfrsf14_mut)
  tips$cd274 <- get_tip_annot(cd274_mut)
  tips$loh_9p <- get_tip_annot(loh_9p == "1")

  # plot
  adj_unit <- -0.016 * max(ape::node.depth.edgelength(p_tree))
  par(family = "sans")
  plot(p_tree, cex = 0.6)
  ape::axisPhylo(side = 1, backward = FALSE)

  # add mutation tip annotations
  ape::tiplabels(pch = 19, tip = tips$activ_b, col = "#55c552", cex = 1, adj = adj_unit * 1)
  ape::tiplabels(pch = 66, tip = tips$activ_b, col = "white", cex = 0.5, adj = adj_unit * 1)
  ape::tiplabels(pch = 19, tip = tips$tnfrsf14, col = "#73065d", cex = 1, adj = adj_unit * 2)
  ape::tiplabels(pch = 84, tip = tips$tnfrsf14, col = "white", cex = 0.5, adj = adj_unit * 2)
  ape::tiplabels(pch = 19, tip = tips$loh_1p, col = "darkblue", cex = 1, adj = adj_unit * 3)
  ape::tiplabels(pch = 49, tip = tips$loh_1p, col = "white", cex = 0.5, adj = adj_unit * 3)
  ape::tiplabels(pch = 19, tip = tips$cd274, col = "#b22c07", cex = 1, adj = adj_unit * 4)
  ape::tiplabels(pch = 67, tip = tips$cd274, col = "white", cex = 0.5, adj = adj_unit * 4)
  ape::tiplabels(pch = 19, tip = tips$loh_9p, col = "#f0a500", cex = 1, adj = adj_unit * 5)
  ape::tiplabels(pch = 57, tip = tips$loh_9p, col = "white", cex = 0.5, adj = adj_unit * 5)
}
```

```{r plot_dna_tree, fig.height = 9}
plot_tree(tree$dna)
```

```{r plot_dnahyb_tree, fig.height = 15, fig.width = 10}
plot_tree(tree$dnahyb)
```

# Differential gene expression analysis

```{r expr_mat}
# get cells and their tnfrsf14 status
cells <-
  ss_annot %>%
  distinct(cell_id, celltype_VDJ_SHM_CSR, loh_1p) %>%
  left_join(
    genos_filter %>%
      filter(gene == "TNFRSF14", !impact %in% c("Synonymous", "Non-coding")) %>%
      distinct(cell_id) %>%
      mutate(TNFRSF14_mut = TRUE)) %>%
  mutate(sample_id = cell_id,
         condition = factor(as.numeric(!is.na(TNFRSF14_mut)),
                            levels = c("0", "1"))) %>%
  filter(celltype_VDJ_SHM_CSR == "activated B")

# get expression in those cells
expr <-
  "out/resolveome/basejumper/bj-expression/rna/PD63118_run/secondary_analyses/quantification_htseq/matrix_gene_counts_starhtseq.txt" %>%
  readr::read_tsv() %>%
  tidyr::separate_wider_delim(cols = "gene_id_gene_symbol", delim = "_",
                              names = c("gene_id", "gene_symbol")) %>%
  rename_with(~ stringr::str_replace(., "_rna_.*", ""))
expr_counts <-
  expr %>%
  select(gene_id, gene_symbol, any_of(cells$sample_id)) %>%
  column_to_rownames(var = "gene_id") %>%
  select(-gene_symbol)

# ensure order of sample IDs matches
cells <-
  cells %>%
  filter(sample_id %in% colnames(expr_counts)) %>%
  arrange(match(sample_id, colnames(expr_counts)))

# convert to a DESeq2-compatible format
dds <-
  DESeqDataSetFromMatrix(countData = as.matrix(expr_counts),
                         colData = cells, design = ~ condition)
dds$condition <- relevel(dds$condition, ref = "0")

# run DESeq2 normalization and differential expression analysis
dds <- DESeq(dds)

# to visualize normalized counts for a specific gene
normalised_counts <- counts(dds, normalized = TRUE)

# extract results
res <-
  results(dds, contrast = c("condition", "1", "0")) %>%
  as_tibble(rownames = "gene") %>%
  arrange(padj) %>%
  left_join(expr %>% select(gene = gene_id, gene_symbol)) %>%
  mutate(direction = ifelse(log2FoldChange > 0, "upregulated", "downregulated"))
significant_genes <-
  res %>%
  filter(padj < 0.05)
print(significant_genes)
```

```{r volcano, fig.height = 7, fig.width = 10}
# plot volcano
labs <-
  res %>%
  group_by(direction) %>%
  top_n(10, wt = -log10(padj))
res %>%
  filter(!is.na(padj)) %>%
  mutate(sig_dir = case_when(padj < 0.05 & direction == "upregulated" ~ "upregulated",
                             padj < 0.05 & direction == "downregulated" ~ "downregulated",
                             TRUE ~ "not significant")) %>%
  ggplot(aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = sig_dir, shape = padj < 0.05), alpha = 0.6) +
  ggrepel::geom_text_repel(data = labs, aes(label = gene_symbol), size = 3,
                           vjust = -0.5) +
  scale_color_manual(values = c("upregulated" = "darkred",
                                "downregulated" = "darkblue",
                                "not significant" = "black")) +
  scale_shape_manual(values = c(1, 19)) +
  theme_minimal() +
  labs(title = "Differentially expressed genes",
       subtitle = "TNFRSF14 mutant vs WT activated B cells",
       x = "log2 fold change",
       y = "-log10 adjusted p-value")
```

```{r heatmap, fig.height = 12, fig.width = 10}
# plot heatmap
# Get top 50 differentially expressed genes
top_genes <-
  significant_genes %>%
  group_by(direction) %>%
  top_n(50, wt = -log10(padj)) %>%
  {split(., .$direction)}

# function: plot heatmap
plot_heatmap <- function(top_genes_i, counts_i, p_title = "Top 50 DGEs",
                         annotations = c()) {
  # Extract their expression data
  expr_top <- expr_counts[top_genes_i$gene, arrange(cells, condition)$sample_id]
  rownames(expr_top) <- top_genes_i$gene_symbol

  # Create a column annotation for TNFRSF14 status
  annotation_col <-
    cells %>%
    transmute(condition = factor(condition), celltype_VDJ_SHM_CSR, sample_id, loh_1p) %>%
    tibble::column_to_rownames(var = "sample_id")

  # Define color scheme
  ann_colors <- list(condition = c("0" = "blue", "1" = "red"))

  # Scale and visualize heatmap
  pheatmap::pheatmap(log2(expr_top + 1),
           scale = "row", cluster_cols = FALSE,
           clustering_distance_rows = "correlation",
           annotation_col = annotation_col,
           annotation_colors = ann_colors,
           main = p_title)
}

# plot up/downregulated genes
plot_heatmap(top_genes$upregulated, expr_counts, p_title = "Top 50 upregulated DEGs")
plot_heatmap(top_genes$downregulated, expr_counts, p_title = "Top 50 downregulated DEGs")
```

```{r dgea_stats}
dgea_counts <-
  cells %>% dplyr::count(condition) %>% {split(.$n, .$condition)}
updown_counts <-
  significant_genes %>% dplyr::count(direction) %>% {split(.$n, .$direction)}
```

Differential gene expression analysis was performed between TNFRSF14-mutant 
(n = `r dgea_counts[["1"]]`) and wildtype (n = `r dgea_counts[["0"]]`) activated
B cells. `r updown_counts$upregulated` genes were significantly upregulated and
`r updown_counts$downregulated` were significantly downregulated [see figure X].

# Pathway enrichment analysis

Run pathway enrichment analysis on the cells.

```{r path_enrich}
# function: run GO and KEGG pathway enrichment on a list of genes
run_pathway_enrichment <- function(genes) {
  # map Ensembl to Entrez IDs for enrichment
  sig_entrez <-
    genes %>%
    tools::file_path_sans_ext() %>%
    bitr(fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = org.Hs.eg.db) %>%
    left_join(
      significant_genes %>% transmute(ENSEMBL = tools::file_path_sans_ext(gene),
                                      gene_symbol = gene_symbol))

  # run GO enrichment
  ego <- enrichGO(gene = sig_entrez$ENTREZID,
                  OrgDb = org.Hs.eg.db,
                  keyType = "ENTREZID",
                  ont = "BP",
                  pAdjustMethod = "BH",
                  pvalueCutoff = 0.05,
                  qvalueCutoff = 0.1)

  # run KEGG enrichment
  ekegg <- enrichKEGG(gene = sig_entrez$ENTREZID,
                      organism = "hsa",
                      pvalueCutoff = 0.05)

  # return
  list(ego = ego, ekegg = ekegg)
}

# run for upreg and downreg
pe_upreg <-
  significant_genes %>%
  filter(log2FoldChange > 1) %>%
  pull(gene) %>%
  run_pathway_enrichment()
pe_downreg <-
  significant_genes %>%
  filter(log2FoldChange < -1) %>%
  pull(gene) %>%
  run_pathway_enrichment()
```

```{r plot_up_pathway_enrich, fig.height = 7}
# visualize top results
p1 <- barplot(pe_upreg$ego, showCategory = 10, title = "GO Biological Process Enrichment of upregulated genes")
p2 <- barplot(pe_downreg$ego, showCategory = 10, title = "GO Biological Process Enrichment of downregulated genes")
(p1 / p2) + plot_layout(heights = c(1, 10), guides = "collect")
```

```{r pathway_genes}
# map Ensembl to Entrez IDs for enrichment
sig_entrez <-
  significant_genes$gene %>%
  tools::file_path_sans_ext() %>%
  bitr(fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = org.Hs.eg.db) %>%
  left_join(
    significant_genes %>% transmute(ENSEMBL = tools::file_path_sans_ext(gene),
                                    gene_symbol = gene_symbol))

# upreg genes involved from GO output
pe_upreg$ego@result %>%
  dplyr::rename(ENTREZID = geneID) %>%
  filter(p.adjust < 0.05) %>%
  tidyr::separate_longer_delim(ENTREZID, delim = "/") %>%
  left_join(sig_entrez) %>%
  group_by(Description, GeneRatio, FoldEnrichment, p.adjust) %>%
  summarise(genes = paste(gene_symbol, collapse = ", ")) %>%
  knitr::kable()

# downreg genes involved from GO output
pe_downreg$ego@result %>%
  dplyr::rename(ENTREZID = geneID) %>%
  filter(p.adjust < 0.05) %>%
  tidyr::separate_longer_delim(ENTREZID, delim = "/") %>%
  left_join(sig_entrez) %>%
  group_by(Description, GeneRatio, FoldEnrichment, p.adjust) %>%
  summarise(genes = paste(gene_symbol, collapse = ", ")) %>%
  knitr::kable()

# upreg genes involved from KEGG output
pe_upreg$ekegg@result %>%
  dplyr::rename(ENTREZID = geneID) %>%
  filter(p.adjust < 0.05) %>%
  tidyr::separate_longer_delim(ENTREZID, delim = "/") %>%
  left_join(sig_entrez) %>%
  group_by(Description, GeneRatio, FoldEnrichment, p.adjust) %>%
  summarise(genes = paste(gene_symbol, collapse = ", ")) %>%
  knitr::kable()

# downreg genes involved from KEGG output
pe_downreg$ekegg@result %>%
  dplyr::rename(ENTREZID = geneID) %>%
  filter(p.adjust < 0.05) %>%
  tidyr::separate_longer_delim(ENTREZID, delim = "/") %>%
  left_join(sig_entrez) %>%
  group_by(Description, GeneRatio, FoldEnrichment, p.adjust) %>%
  summarise(genes = paste(gene_symbol, collapse = ", ")) %>%
  knitr::kable()
```

```{r pathway_stats}
pe_upreg$ego
pe_downreg$ego
``` 
