---
title: "`r params$title`"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: true
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
params:
  title: null
  cache_path: null
  rerun: false
---

```{r setup, include = FALSE}
# set knitr options
knitr::opts_knit$set(root.dir = "../..")
knitr::opts_chunk$set(
  warning = FALSE,
  message = FALSE,
  dpi = 300,
  cache.path = params$cache_path,
  fig.align = "center"
)

# set ggplot presets
ggplot2::theme_set(ggplot2::theme_bw())
```

```{r setup2, include = FALSE}
# libraries
library(Matrix)
library(dplyr)
library(GenomicRanges)
library(rtracklayer)
library(ggplot2)
library(data.table)
library(patchwork)
source("bin/utils.R")

# dirs
vartrix_data_dir <-
  file.path(Sys.getenv("LUSTRE_TEAM"),
            "projects/hashimoto_thyroiditis/data/vartrix/")
vartrix_out_dir <-
  file.path(Sys.getenv("LUSTRE_TEAM"),
            "projects/hashimoto_thyroiditis/out/vartrix/")
```

VarTrix was run on the 10X and PacBio Trencadis-Seq BAMs to genotype 1p SNPs
that have been phased, to look for cells with evidence of 1p loss. We will
now use the output to call the haplotypes of 1p variants in the cells and to
determine whether there is sufficient evidence to call 1p loss.

First, we load the celltype annotations for this dataset.

```{r load_celltype_annotations}
# load annotated seurat object
seu <-
  list.files("out/trencadis-seq/seurat/ncells_3_nfeature_500-2000_ncount_1000-15000_percent_mt_5/nf-trencadis-seq_thyroid_annotate_celltypes_cache/html/",
             pattern = "seu_annotated.*\\.rds",
             full.names = TRUE) %>%
  readRDS()

# plot umap
DimPlot(seu, group.by = "celltype") + coord_equal()
```

Next, we load the VarTrix output for the 10X and PacBio Trencadis-Seq BAMs.

```{r load_data}
# list of vartrix outputs and barcode files
vartrix_files <-
  list.files(vartrix_out_dir) %>%
  purrr::set_names() %>%
  purrr::map(function(kit) {
    list.files(file.path(vartrix_out_dir, kit, "filtered")) %>%
      purrr::set_names() %>%
      purrr::map(function(sample) {
        file.path(vartrix_out_dir, kit, "filtered", sample)
      })
  })

# get matrices
mats <- list()
purrr::walk2(names(vartrix_files), vartrix_files, function(kit_i, kit) {
  purrr::walk2(names(kit), kit, function(sample_i, sample_dir) {

    # create id (kit x sample)
    id_i <- paste0(kit_i, "_", sample_i)

    # load barcodes
    barcodes <- readLines(file.path(sample_dir, "barcodes.tsv"))

    # convert pacbio barcodes
    if (kit_i == "PB") {
      barcodes <- rev_comp(barcodes) %>% paste0("-1")
    }

    # load variants and correct positions (+1)
    vars <- readLines(file.path(sample_dir, "variants.txt"))
    chrs <- gsub("_.*", "", vars) %>% gsub("chr", "", .)
    positions_plus_1 <- gsub(".*_", "", vars) %>% as.numeric() %>% {. + 1}
    vars_plus_1 <- paste0(chrs, "_", positions_plus_1)

    # load matrices
    mats[[id_i]] <<-
      c("alt", "ref") %>%
      purrr::set_names() %>%
      purrr::map(function(x) {

        # load matrix
        mat <- readMM(file.path(sample_dir, paste0(x, ".mtx")))
        dimnames(mat) <- list(vars_plus_1, barcodes)
        mat

      })
  })
})
```

Now, we load the SNPs and order them in the same way as in the matrices.

```{r load_snps}
Sys.setenv("VROOM_CONNECTION_SIZE" = "5000000")
snps <-
  file.path(
    Sys.getenv("LUSTRE_TEAM"), "projects/hashimoto_thyroiditis/data",
    "vartrix/snps/PD63118b_lo0044.v1.caveman_c.snps.1p_nochr.vcf.gz") %>%
  gzfile() %>%
  readr::read_tsv(comment = "##") %>%
  transmute(chr = as.character(`#CHROM`), pos = POS, ref = REF, alt = ALT)
mat_snps <-
  tibble::tibble(chr_pos = rownames(mats[[1]][[1]])) %>%
  tidyr::separate_wider_delim(cols = "chr_pos", delim = "_",
                              names = c("chr", "pos"), cols_remove = FALSE) %>%
  mutate(pos = as.numeric(pos)) %>%
  left_join(snps)
```

Now, we subset the matrices to only annotated cells.

```{r filter_matrices}
# get all barcodes
barcodes <-
  purrr::map2(names(mats), mats, function(id_i, i) {
    tibble::tibble(barcode = colnames(i[[1]])) %>%
      mutate(!!id_i := TRUE)
  }) %>%
  Reduce(f = dplyr::full_join) %>%
  # annotated barcodes
  mutate(annotated = barcode %in% colnames(seu))

# we only consider cells that are annotated
barcodes_filter <-
  barcodes %>%
  filter(annotated) %>%
  pull(barcode)

# subset matrices
mats_filter <-
  purrr::map(mats, function(i) {
    purrr::map(i, function(m) {
      barcodes <- colnames(m)
      keep_cols <- which(barcodes %in% barcodes_filter)
      m[, keep_cols, drop = FALSE]
    })
  })

# how many barcodes are shared between the two samples and are annotated?
barcodes %>%
  group_by(across(-barcode)) %>%
  dplyr::count() %>%
  knitr::kable()
```

Now, we generate some summary statistics fo the genotyping efficiency of each 
run.

```{r summary_stats}
mats_filter %>%
  purrr::map(function(i) {
    cell_cov <- Matrix::colSums(i$alt) + Matrix::colSums(i$ref)
    sum(cell_cov > 0)
  })
```

Now, we align the matrices and sum them.

```{r sum_matrices}
# align matrices
mats_aligned <- list(ref = list(), alt = list())
purrr::walk2(names(mats_filter), mats_filter, function(sample, x) {
  purrr::walk2(names(x), x, function(type, m) {
    # reorder and fill in missing barcodes
    missing <- setdiff(barcodes_filter, colnames(m))
    if (length(missing) > 0) {
      zero_mat <- Matrix(0, nrow = nrow(m), ncol = length(missing), sparse = TRUE)
      colnames(zero_mat) <- missing
      rownames(zero_mat) <- rownames(m)
      m <- cbind(m, zero_mat)
    }
    # reorder to consistent barcode order
    mats_aligned[[type]][[sample]] <<- m[, barcodes_filter]
  })
})

# sum the matrices
mats_summed <-
  mats_aligned %>%
  purrr::map(~ Reduce(`+`, .x))

# calculate vafs and total depths
mat_calc <-
  list(alt_vaf = mats_summed$alt / (mats_summed$alt + mats_summed$ref),
       alt_depth = mats_summed$alt,
       ref_depth = mats_summed$ref,
       total_depth = mats_summed$alt + mats_summed$ref)

# we subset to positions with any coverage in any cell (informative)
mat_calc <-
  purrr::map(mat_calc, ~ .x[rowSums(mat_calc$total_depth) > 0, ])
```

Now, we can call the haplotypes of the 1p variants in the cells.

```{r call_haplotypes}
# load grch37 snps
phased_snps_grch37 <-
  readRDS("out/phase_snps/phase_snps_cache/plate3_wellA2_phased_snps.rds") %>%
  dplyr::rename(pos_grch37 = pos)

# lift over to grch38
phased_snps <-
  phased_snps_grch37 %>%
  {GRanges(
    seqnames = paste0("chr", .$chr),
    ranges = IRanges(start = .$pos_grch37, end = .$pos_grch37),
    pos_grch37 = .$pos_grch37)} %>%
  liftOver(import.chain("../../reference/liftover/hg19ToHg38.over.chain")) %>%
  unlist() %>%
  tibble::as_tibble() %>%
  transmute(chr = gsub("chr", "", seqnames), pos = start, pos_grch37,
            chr_pos = paste0(chr, "_", pos)) %>%
  left_join(phased_snps_grch37) %>%
  # phased snps only + only those in the matrix
  filter(!is.na(hap_A_type),
         chr_pos %in% rownames(mat_calc$alt_vaf))

# subset vaf matrix to phased snps
mat_calc_phased <- purrr::map(mat_calc, ~ .x[phased_snps$chr_pos, ])

# update phased snps to match matrix order
phased_snps <-
  left_join(tibble::tibble(chr_pos = rownames(mat_calc_phased$alt_vaf)),
            phased_snps)

# flip VAF (ie 1 - VAF) where hap A == "ref", to get VAFs of hap A for all cells
mat_calc_phased$alt_vaf_A <- mat_calc_phased$alt_vaf
mat_calc_phased$alt_vaf_A[phased_snps$hap_A_type == "ref", ] <-
  1 - mat_calc_phased$alt_vaf_A[phased_snps$hap_A_type == "ref", ]
```

We look at those sites with any reads.

```{r mat_long}
mat_long <-
  xfun::cache_rds({
    mat_calc_phased %>%
      purrr::map(function(m) {

        # get all possible combinations
        all_combinations <- expand.grid(
          chr_pos = rownames(m),
          barcode = colnames(m),
          stringsAsFactors = FALSE
        )

        # get non-zero entries from sparse matrix
        mat_triplet <- as(m, "TsparseMatrix")
        nonzero_entries <- data.table(
          chr_pos = rownames(m)[mat_triplet@i + 1],
          barcode = colnames(m)[mat_triplet@j + 1],
          value = mat_triplet@x
        )

        # merge to include zeros
        all_combinations %>%
          left_join(nonzero_entries, by = c("chr_pos", "barcode")) %>%
          mutate(value = ifelse(is.na(value), 0, value))

      }) %>%
      bind_rows(.id = "name") %>%
      tidyr::pivot_wider() %>%
      # filter out sites with no reads
      filter(total_depth > 0) %>%
      left_join(phased_snps) %>%
      # order by index
      group_by(barcode) %>%
      arrange(chr, pos) %>%
      mutate(i = row_number(), mean_alt_vaf_A = mean(alt_vaf_A), n = n()) %>%
      # add annotations and factor
      left_join(tibble::enframe(seu$celltype, name = "barcode",
                                value = "celltype")) %>%
      add_count(celltype, name = "n_ct") %>%
      mutate(celltype = forcats::fct_reorder(celltype, -n_ct)) %>%
      # factor the barcodes
      ungroup() %>%
      arrange(celltype, mean_alt_vaf_A, n) %>%
      mutate(barcode = forcats::fct_inorder(barcode)) %>%
      # factor the celltypes
      group_by(celltype) %>%
      mutate(n_cells = n_distinct(barcode)) %>%
      ungroup() %>%
      arrange(-n_cells) %>%
      mutate(celltype = forcats::fct_inorder(celltype))
  }, file = "mat_long.rds", rerun = params$rerun)
```

We plot the distribution of VAFs and depths for the cells.

```{r plot_vaf_dist}
n_unique_muts <-
  mat_long %>%
  dplyr::distinct(chr, pos, ref, alt) %>%
  nrow() %>%
  format(big.mark = ",", scientific = FALSE)
n_total_muts <-
  mat_long %>%
  nrow() %>%
  format(big.mark = ",", scientific = FALSE)

# plot
mat_long %>%
  dplyr::mutate(
    alt_vaf_bin = cut(alt_vaf, breaks = seq(0, 1, length.out = 30 + 1),
                      include.lowest = TRUE),
    alt_vaf_min = stringr::str_extract(alt_vaf_bin, "[\\d\\.]+") %>% as.numeric(),
    alt_vaf_high = as.numeric(stringr::str_extract(alt_vaf_bin, "(?<=,)\\s*[\\d\\.]+")),
    alt_vaf_mid = (alt_vaf_min + alt_vaf_high) / 2) %>%
  dplyr::count(alt_vaf_bin, alt_vaf_mid, total_depth) %>%
  ggplot(aes(x = alt_vaf_mid, y = n, fill = total_depth)) +
  geom_col() +
  viridis::scale_fill_viridis() +
  theme_classic() +
  labs(title = "distribution of VAFs",
                subtitle = paste(n_total_muts, "total mutations,",
                                 n_unique_muts, "unique mutations"),
        x = "VAF bin", y = "n mutations") +
  lims(x = c(0, 1)) +
  scale_y_continuous(expand = c(0, 0))
```

We plot their haplotype A VAFs, sorted by mean haplotype A VAF.

```{r plot_vafs, fig.height = 8}
# plot
c("alt_vaf_A", "alt_vaf", "total_depth") %>%
  purrr::walk(function(type) {
    p1 <-
      mat_long %>%
      ggplot(aes(x = 1, y = barcode, fill = mean_alt_vaf_A)) +
      geom_raster(interpolate = FALSE) +
      viridis::scale_fill_viridis(option = "A") +
      theme(axis.title.y = element_blank(), axis.text.y = element_blank(),
            strip.text.y.left = element_text(angle = 0),
            axis.ticks.y = element_blank(), axis.ticks.x = element_blank(),
            axis.title.x = element_blank(), axis.text.x = element_blank()) +
      scale_x_continuous(expand = c(0, 0)) +
      facet_grid(celltype ~ ., scales = "free", space = "free", switch = "y")
    p2 <-
      mat_long %>%
      ggplot(aes(x = i, y = barcode, fill = !!sym(type))) +
      geom_raster(interpolate = FALSE) +
      viridis::scale_fill_viridis() +
      theme(axis.title.y = element_blank(), axis.text.y = element_blank(),
            axis.ticks.y = element_blank(), strip.background = element_blank(),
            strip.text.y.left = element_blank()) +
      scale_x_continuous(expand = c(0, 0)) +
      facet_grid(celltype ~ ., scales = "free", space = "free", switch = "y") +
      labs(x = "n SNPs", fill = type)
    p <- (p1 | p2) + plot_layout(guides = "collect", widths = c(1, 8))
    print(p)
  })
```

We plot their haplotype A VAFs, sorted by leading runs of 1s and 0s.

```{r plot_vaf_runs, fig.height = 8}
# compute number of *leading* alt_vaf_A == 1/0 per barcode
leading_runs <-
  mat_long %>%
  arrange(barcode, i) %>%
  group_by(barcode) %>%
  mutate(
    alt1 = !is.na(alt_vaf_A) & alt_vaf_A == 1,
    alt0 = !is.na(alt_vaf_A) & alt_vaf_A == 0
  ) %>%
  summarise(
    leading_run_1 = {
      r <- rle(alt1)
      if (length(r$values) > 0 && isTRUE(r$values[1])) r$lengths[1] else 0
    },
    leading_run_0 = {
      r <- rle(alt0)
      if (length(r$values) > 0 && isTRUE(r$values[1])) r$lengths[1] else 0
    },
    .groups = "drop"
  )

# join back and arrange by both
mat_long_ordered <-
  mat_long %>%
  left_join(leading_runs, by = "barcode") %>%
  arrange(desc(leading_run_1), desc(leading_run_0), barcode, i) %>%
  mutate(barcode = forcats::fct_inorder(barcode))

# plot
c("alt_vaf_A", "alt_vaf", "total_depth") %>%
  purrr::walk(function(type) {
    p <-
      mat_long_ordered %>%
      ggplot(aes(x = i, y = barcode, fill = !!sym(type))) +
      geom_raster(interpolate = FALSE) +
      viridis::scale_fill_viridis() +
      theme(axis.title.y = element_blank(), axis.text.y = element_blank(),
            axis.ticks.y = element_blank(), strip.placement = "outside",
            strip.text.y.left = element_text(angle = 0)) +
      scale_x_continuous(expand = c(0, 0)) +
      facet_grid(celltype ~ ., scales = "free", space = "free", switch = "y") +
      labs(x = "n SNPs", fill = type)
    print(p)
  })
```

We plot the distribution of these reads along 1p.

```{r plot_vaf_dist_1p, fig.height = 8}
# relevant boundaries
chr1p_end <- 122026459
tnfrsf14_start <- 2556371

# plot the positions of 10X-genotyped snps along chr1p
p1 <-
  mat_long %>%
  mutate(`pos (Mb)` = pos / 1e6) %>%
  ggplot(aes(x = `pos (Mb)`, y = barcode, colour = alt_vaf_A)) +
  geom_point(size = 0.3) +
  viridis::scale_colour_viridis() +
  theme(axis.text.y = element_blank(), axis.title.y = element_blank(),
        axis.ticks.y = element_blank(), strip.placement = "outside",
        strip.text.y.left = element_text(angle = 0)) +
  facet_grid(celltype ~ ., scales = "free", space = "free", switch = "y") +
  scale_x_continuous(limits = c(0, chr1p_end / 1e6), expand = c(0, 0))

# plot the bounds of chr1p (1:121500000)
p2 <-
  tibble::tibble(x = "chr1p", `pos (Mb)` = chr1p_end / 1e6) %>%
  ggplot(aes(x = x, y = `pos (Mb)`)) +
  geom_col() +
  geom_hline(yintercept = tnfrsf14_start / 1e6, colour = "green", size = 2) +
  coord_flip() +
  theme(axis.title.y = element_blank()) +
  scale_y_continuous(limits = c(0, chr1p_end / 1e6), expand = c(0, 0))

(p1 / p2) + plot_layout(heights = c(10, 1))
```

How many cells have at least 10 reads across 1p?

```{r count_cells_10_reads}
mat_long %>%
  group_by(barcode, celltype) %>%
  summarise(total_depth_1p = sum(total_depth)) %>%
  ggplot(aes(x = total_depth_1p)) +
  geom_histogram() +
  geom_vline(xintercept = 10) +
  facet_wrap(~ celltype)
```