---
title: "`r params$title`"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: true
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
params:
  title: null
  cache_path: null
  rerun: false
---

```{r setup, include = FALSE}
knitr::opts_knit$set(root.dir = "../..")
knitr::opts_chunk$set(
  warning = FALSE,
  message = FALSE,
  dpi = 300,
  cache.path = params$cache_path,
  fig.align = "center"
)
```

```{r setup2, include = FALSE}
# libraries
library(magrittr)
library(dplyr)
library(ggplot2)
library(readr)
library(tidyr)
library(liftOver)
library(ape)
library(patchwork)

# set ggplot presets
theme_set(theme_classic())

# dirs
seq_dir <- "out/resolveome/sequoia/20250918/"
```

# Parameters

```{r params, echo = FALSE}
params
```

We plot the tree.

```{r tree, fig.width = 10, fig.height = 10}
# load tree
tree <-
  file.path(seq_dir, "Patient_both_tree_with_branch_length.tree") %>%
  ape::read.tree()

# load cell id to new cell id
cell_annots <-
  tibble::tibble(id = tree$tip.label) %>%
  left_join(
    file.path(Sys.getenv("LUSTRE_125"),
              "projects/hashimoto_thyroiditis/out/resolveome/basejumper",
              "bj-somatic-variantcalling/dna/PD63118/samplesheet.csv") %>%
      readr::read_csv() %>%
      transmute(id = biosampleName, cell_id = gsub("_dna.*", "", id)) %>%
      # add new cells ids + additional annotations
      left_join(
        "data/resolveome/manual_inspection/pta_additional_annotation_H1.tsv" %>%
          readr::read_tsv() %>%
          mutate(cell_id_new = as.character(cell_ID), cell_id = well_ID)
      )
  )
```

```{r plot_tree, fig.width = 10, fig.height = 10}
# tree with ids
plot(tree, cex = 0.7)
axisPhylo(side = 1, backward = FALSE)

# tree with ids, equal branch lengths
tree_collapsed <- tree
tree_collapsed$edge.length <- rep(1, nrow(tree_collapsed$edge))
plot(tree_collapsed)
```

We load the mutations.

```{r load_muts}
# load mutations assigned to branches
muts_to_branches <-
  file.path(seq_dir, "Patient_both_assigned_to_branches.txt") %>%
  readr::read_tsv() %>%
  janitor::clean_names()
```

Now we plot the tree with various annotations.

```{r plot_tree_relabel, fig.width = 10, fig.height = 10}
# rename tips
tree_relabel <- tree
tree_relabel$tip.label <- cell_annots$cell_id_new
write.tree(tree_relabel, file.path(seq_dir, "Patient_both_tree_relabelled.tree"))

# plain tree
par(family = "sans")
plot(tree_relabel, cex = 0.7, font = 1)
axisPhylo(side = 1, cex = 0.7, backward = FALSE)

# plain tree with uniform branch lengths
tree_collapsed <- tree_relabel
tree_collapsed$edge.length <- rep(1, nrow(tree_collapsed$edge))
plot(tree_collapsed)

# tree coloured by celltype
tip_cols <- ifelse(cell_annots$celltype_VDJ_recomb == "B cell", "blue",
            ifelse(cell_annots$celltype_VDJ_recomb == "alpha-beta T cell",
                   "red", "grey"))
plot(tree_relabel, tip.color = tip_cols, cex = 0.7, font = 1)
axisPhylo(side = 1, backward = FALSE)
legend("topright",
       legend = c("B cell", "alpha-beta T cell", "not lymphocyte"),
       col    = c("blue", "red", "grey"),
       pch    = 19,
       pt.cex = 1,
       bty    = "n")

# tree coloured by biallelic_CD274
tip_cols <- ifelse(cell_annots$biallelic_CD274 == "TRUE_Compound", "blue",
            ifelse(cell_annots$biallelic_CD274 == "TRUE_LOH", "#3ec83e",
            ifelse(cell_annots$biallelic_CD274 == "FALSE_Het", "red",
            ifelse(cell_annots$biallelic_CD274 == "Unclear - 9p del + intronic",
                   "purple", "grey"))))
plot(tree_relabel, tip.color = tip_cols, cex = 0.7)
axisPhylo(side = 1, backward = FALSE)
legend("topright",
       legend = c("TRUE_Compound", "TRUE_LOH", "FALSE_Het",
                  "Unclear - 9p del + intronic", "WT/-"),
       col    = c("blue", "#3ec83e", "red", "purple", "grey"),
       pch    = 19,
       pt.cex = 1,
       bty    = "n")

# tree coloured by SHM
tip_cols <- ifelse(cell_annots$celltype_SHM == "Mature B cell", "blue",
                   "red")
plot(tree_relabel, tip.color = tip_cols, cex = 0.7)
axisPhylo(side = 1, backward = FALSE)
legend("topright",
       legend = c("Mature B cell", "Not mature B cell"),
       col    = c("blue", "red"),
       pch    = 19,
       pt.cex = 1,
       bty    = "n")
```

Now, we mark driver mutations on the tree.

```{r plot_drivers_on_tree}
# create tree df
tree_df <- as.data.frame(ggtree::fortify(tree_relabel))

# visualise edges and nodes
plot(tree_collapsed, cex = 0.5)
edgelabels(cex = 0.5)
nodelabels(cex = 0.5)

# get shared clades
shared_clades <-
  readr::read_tsv("data/resolveome/shared_clades/shared_clades.tsv")

# get genotyped mutations
muts <-
  # load genotyping
  c("dna", "dnahyb") %>%
  purrr::map(function(seq_type) {
    file.path("out/resolveome/nf-resolveome", seq_type, "PD63118",
              "genotyping/mutations/PD63118_annotated_mutations.tsv") %>%
      readr::read_tsv() %>%
      filter(gene %in% c("TNFRSF14", "CD274"))
  }) %>%
  # combine dna + dnahyb reads
  bind_rows() %>%
  mutate(cell_id = gsub("_dna.*", "", id)) %>%
  group_by(chr, pos, ref, alt, gene, cell_id) %>%
  summarise(alt_depth = sum(alt_depth), total_depth = sum(total_depth)) %>%
  mutate(alt_vaf = alt_depth / total_depth) %>%
  ungroup() %>%
  # annotate new cell ids
  inner_join(cell_annots) %>%
  # positive mutations only
  filter(alt_vaf > 0.2, alt_depth > 1)

muts %>%
  left_join(cell_annots) %>%
  add_count(chr, pos, ref, alt) %>%
  arrange(-n, chr, pos, ref, alt) %>%
  dplyr::select(chr, pos, ref, alt, gene, cell_id_new, alt_depth, total_depth, alt_vaf, clade_ID, subclade_ID, n)

# get mrca edge of each shared mutations
tips_ls <-
  muts %>%
  split(.$gene) %>%
  purrr::map(function(gene_df) {
    gene_df %>%
      group_by(chr, pos, ref, alt) %>%
      group_split() %>%
      purrr::map(~ .x$cell_id_new)
  })

# function: plot mutations on the tree
plot_phylo_muts <- function(tree, tips_ls) {

  # get celltype colours
  tip_cols_legend <-
    c("B cell" = "#1515a4", "alpha-beta T cell" = "#9e0303",
      "not lymphocyte" = "grey")
  tip_cols <- tip_cols_legend[cell_annots$celltype_VDJ_recomb]

  # get gene colours
  if (length(tips_ls) == 1) {
    gene_cols <- "orange"
  } else if (length(tips_ls) == 2) {
    gene_cols <- c("orange", "purple")
  } else {
    gene_cols <- RColorBrewer::brewer.pal(n = length(tips_ls), name = "Set1")
  }
  names(gene_cols) <- names(tips_ls)

  # plot tree
  par(family = "sans")
  plot(tree, cex = 0.7, font = 1, tip.color = tip_cols)
  coords <- get("last_plot.phylo", envir = .PlotPhyloEnv)

  # start list of edges
  edges <- list()

  purrr::walk2(names(tips_ls), tips_ls, function(gene, gene_tips) {

    edges[[gene]] <<- list()

    purrr::walk(gene_tips, function(tips) {

      # get edge before tip or mrca node
      if (length(tips) == 1) {
        mrca_node <- which(tree$tip.label == tips)
        edge <- which(tree$edge[, 2] == mrca_node)
      } else {
        mrca_node <- getMRCA(tree, tips)
        edge <- which(tree$edge[, 2] == mrca_node)

        # make sure edge descends to tips only
        desc_tips <- extract.clade(tree, mrca_node)$tip.label
        if (!all(desc_tips %in% tips)) {
          # if not, then get edge before each tip
          edge <-
            tips %>%
            purrr::map(function(tip) {
              which(tree$edge[, 2] == which(tree$tip.label == tip))
            }) %>%
            unlist()
        }
      }

      # add edge(s) to list of edges
      edges[[gene]] <<- c(edges[[gene]], list(edge))

      for (e in edge) {
        # edge coordinates (parent -> child)
        x0 <- coords$xx[tree$edge[e, 1]]
        y0 <- coords$yy[tree$edge[e, 1]]
        x1 <- coords$xx[tree$edge[e, 2]]
        y1 <- coords$yy[tree$edge[e, 2]]
        x_mid <- (x0 + x1) / 2
        y_mid <- y1

        # add points
        points(x_mid, y_mid, pch = 19, col = gene_cols[gene], cex = 1)
      }

    })

  })

  # check if any edges are mutated in both genes
  # intersect(unlist(edges[[1]]), unlist(edges[[2]]))
  shared_edges <- Reduce(intersect, lapply(edges, unlist))
  for (e in shared_edges) {
    # edge coordinates (parent -> child)
    x0 <- coords$xx[tree$edge[e, 1]]
    y0 <- coords$yy[tree$edge[e, 1]]
    x1 <- coords$xx[tree$edge[e, 2]]
    y1 <- coords$yy[tree$edge[e, 2]]
    x_mid <- (x0 + x1) / 2
    y_mid <- y1

    # add points
    points(x_mid, y_mid, pch = 19, col = "#d2208b", cex = 1)
    gene_cols <- c(gene_cols, "TNFRSF14 + CD274" = "#d2208b")
  }

  # add axis and legend
  axisPhylo(side = 1, backward = FALSE)
  legend("topright",
         legend = c("Driver mutation", names(gene_cols), "Cell type", names(tip_cols_legend)),
         text.font = c(2, rep(3, length(gene_cols)), 2, rep(1, length(tip_cols_legend))),
         col    = c(NA, gene_cols, rep("black", length(tip_cols_legend) + 1)),
         pch    = c(NA, rep(19, length(gene_cols)), NA, rep(95, length(tip_cols_legend))),
         pt.cex = 1,
         cex = 0.7,
         bty    = "n",
         text.col = c(rep("black", length(gene_cols) + 2),
                      tip_cols_legend))

}

pdf("test.pdf", width = 10, height = 10)
plot_phylo_muts(tree_relabel, tips_ls)
plot_phylo_muts(tree_collapsed, tips_ls)
dev.off()

# plot tree
pdf("test.pdf", width = 10, height = 10)
plot(tree_relabel, label.offset = 0.01 * max(tree_df$x))

# for each branch, draw rectangles showing signature proportions
for (edge in unique(muts_to_branches$branch)) {
  branch <- 8
  x_end <- tree_df$x[n]
  x_start <- tree_df$x[tree_df$parent[n]]
  x_intv <- x_end - x_start
  y <- node.height(tree_relabel)[n]
  tipnum <- sum(tree_df$isTip)

  # stack signature exposures proportionally along branch length
  x_end <- x_start + x_intv / 2
  rect(ybottom = y - min(0.015 * tipnum, 0.3),
       ytop = y + min(0.015 * tipnum, 0.3),
       xleft = x_start, xright = x_end, col = all_cols[s], lwd = 0.25)
}
```