---
title: "Investigate chromosomal dropout in PD63118 in run 49882"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: false
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
---

```{r setup, include = F, message = F, warning = F, echo = F}
# rmarkdown::render('reports/investigate_chr_dropout_PD63118_49882.Rmd', output_file = 'investigate_chr_dropout_PD63118_49882.html', output_dir = 'reports/')

# chunk options
knitr::opts_knit$set(root.dir = "../")
knitr::opts_chunk$set(warning = FALSE,
                      dpi = 300,
                      message = FALSE)

# libraries
library(magrittr)
library(ggplot2)

# pattient
curr_pdid <- "PD63118"
```

We are interested in investigating chromosomal dropout in patient `r curr_pdid`
in run 49882.

```{r}
# samplesheet
ss <-
  readr::read_tsv("data/chr_dropout/PD63118.tsv")
knitr::kable(ss)
```

We can check how cells with apparent chromosomal dropout look in mutation VAF 
heatmap.

First, we load the genotyping data for the CaVEMAN SNPs.

```{r}
geno <-
  "ls out/resolveome/nf-resolveome/PD63118/PD63118_49900/*/genotyping/*_genotyped_mutations.tsv" %>%
  system(intern = TRUE) %>%
  {setNames(., stringr::str_split_i(., "/", 5))} %>%
  purrr::map(readr::read_tsv) %>%
  dplyr::bind_rows(.id = "id")

# prepare data
p_dat <-
  geno %>%
  dplyr::filter(source == "caveman_snps") %>%
  dplyr::left_join(ss) %>%
  dplyr::mutate(mut_id = paste(chr, pos, ref, alt, sep = "_")) %>%
  # count number of cells with each mutation
  dplyr::add_count(mut_id, name = "n_cells_w_mut")
```

Then, we plot a heatmap of SNP VAFs across cells, with cells with chromosomal
dropout annotated, hierarchically clustered.

```{r heatmap_vaf, fig.width = 20, fig.height = 20}
# reshape data for heatmap
p_source <- "caveman_snps"
heatmap_data <-
  p_dat %>%
  reshape2::dcast(mut_id + n_cells_w_mut ~ id, value.var = "mut_vaf")
rownames(heatmap_data) <- heatmap_data$mut_id
heatmap_matrix <- as.matrix(heatmap_data[, -c(1:3)]) # drop annotations
n_total_muts <- dplyr::n_distinct(p_dat$mut_id)
n_total_cells <- dplyr::n_distinct(p_dat$id)

# calculate the number of mutations per column (id)
annotation_col <-
  p_dat %>%
  dplyr::count(id, name = "n_muts") %>%
  dplyr::left_join(ss %>% dplyr::transmute(id, as.numeric(chr_dropout))) %>%
  tibble::column_to_rownames("id")

# replace NAs with 0
heatmap_matrix[is.na(heatmap_matrix)] <- 0

# prepare annotations for columns
annotation_row <- data.frame(n_cells_w_mut = heatmap_data$n_cells_w_mut)
rownames(annotation_row) <- heatmap_data$mut_id

# plot
pdf("test.pdf", width = 20, height = 20)
pheatmap::pheatmap(
  mat = heatmap_matrix,
  cluster_rows = TRUE,
  cluster_cols = TRUE,
  annotation_row = annotation_row,
  annotation_col = annotation_col,
  color = colorRampPalette(c("white", "blue"))(50),
  main = paste(p_source, "- VAF heatmap\n",
               n_total_cells, "cells,", n_total_muts, "mutations")
)
dev.off()
```

Then, we plot a heatmap of SNP VAFs across cells, with cells with chromosomal
dropout annotated, ordered by position along the chromosome to visualise the
large chromosomal events as allelic dropout.

```{r heatmap_vaf, fig.width = 20, fig.height = 20}
# reshape data for heatmap
heatmap_data <-
  p_dat %>%
  reshape2::dcast(mut_id + n_cells_w_mut ~ id, value.var = "mut_vaf")
rownames(heatmap_data) <- heatmap_data$mut_id
heatmap_matrix <- as.matrix(heatmap_data[, -c(1:3)]) # drop annotations
n_total_muts <- dplyr::n_distinct(p_dat$mut_id)
n_total_cells <- dplyr::n_distinct(p_dat$id)

# calculate the number of mutations per column (id)
annotation_col <-
  p_dat %>%
  dplyr::count(id, name = "n_muts") %>%
  dplyr::left_join(ss %>% dplyr::transmute(id, as.numeric(chr_dropout))) %>%
  tibble::column_to_rownames("id")

# replace NAs with 0
heatmap_matrix[is.na(heatmap_matrix)] <- 0

# prepare annotations for columns
annotation_row <- data.frame(n_cells_w_mut = heatmap_data$n_cells_w_mut)
rownames(annotation_row) <- heatmap_data$mut_id

# plot
pheatmap::pheatmap(
  mat = heatmap_matrix,
  cluster_rows = TRUE,
  cluster_cols = TRUE,
  annotation_row = annotation_row,
  annotation_col = annotation_col,
  color = colorRampPalette(c("white", "blue"))(50),
  main = paste(p_source, "- VAF heatmap\n",
               n_total_cells, "cells,", n_total_muts, "mutations")
)
```