---
title: "Genotyping mutations in PTA"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: false
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
---

```{r setup, include = F, message = F, warning = F, echo = F}
# rmarkdown::render('src/03c_analyse_genotyping.Rmd', output_file = 'analyse_genotyping.html', output_dir = 'out/genotyping/')

# chunk options
knitr::opts_knit$set(root.dir = "../")
knitr::opts_chunk$set(warning = FALSE,
                      dpi = 300,
                      message = FALSE)

# libraries
library(magrittr)
library(ggplot2)

# pattient
curr_pdid <- "PD63118"
```

Here, we are analysing the mutations genotyped in PTA data. We want to 
investigate genotyping efficiency, VAF distribution, and allelic imbalance.

For patient `r curr_pdid`, we have genotyped mutations from two sources:

- Mutations from targeted immune and whole exome Nanoseq

- Heterozygous, common SNPs from `CaVEMan` output

We load the mutations from Nanoseq...

```{r load_mutations}
# get mutations from exome and targeted nanoseq
nanoseq_muts <-
  readr::read_tsv("data/nanoseq/samplesheet.tsv") %>%
  purrr::pmap(function(muts_file, seq_type) {
    readr::read_tsv(muts_file) %>%
      dplyr::mutate(seq_type = seq_type)
  }) %>%
  dplyr::bind_rows() %>%
  dplyr::mutate(pdid = stringr::str_sub(sampleID, 1, 7)) %>%
  dplyr::filter(pdid == curr_pdid) %>%
  dplyr::mutate(mut_id = paste(chr, pos, ref, mut, sep = "-"))
nanoseq_muts
```

We load the common SNPs from CaVEMan...

```{r load_snps}
# get common snps from caveman
caveman_snps <-
  readr::read_tsv("out/genotyping/caveman_snps.tsv") %>%
  dplyr::mutate(mut_id = paste(chr, pos, ref, mut, sep = "-"))
caveman_snps
```

There are `r dplyr::n_distinct(nanoseq_muts$mut_id)` mutations from Nanoseq 
and `r dplyr::n_distinct(caveman_snps$mut_id)` SNPs from CaVEMan.

Next, we load genotyping outputs and calculate VAFs...

```{r load_geno}
muts_and_snps <-
  c("caveman_snps", "nanoseq_mutations") %>%
  purrr::set_names() %>%
  purrr::map(function(x) {
    list.files(paste0("out/genotyping/", x, "_genotyped_by_chr/high_qual/"),
               pattern = ".tsv$", full.names = TRUE) %>%
      purrr::map(~ readr::read_tsv(.x) %>% dplyr::mutate(chr = as.character(chr))) %>%
      dplyr::bind_rows() %>%
      dplyr::mutate(mut_vaf = mut_depth / total_depth,
                    mut_id = paste(chr, pos, ref, mut, sep = "-")) 
  }) %>%
  dplyr::bind_rows(.id = "source")
```

The median total depth at each position was
`r median(muts_and_snps$total_depth)`.

The median mutant depth at mutated positions (ie. `mut_depth > 0`) was
`r median(dplyr::filter(muts_and_snps, mut_depth > 0)$mut_depth)`.

How many mutations and SNPs were genotyped?

```{r n_genotyped}
muts_and_snps %>%
  dplyr::mutate(genotyped = as.numeric(mut_depth) > 0) %>%
  dplyr::group_by() %>%
  dplyr::summarise(n = dplyr::n_distinct(mut_id),
                   n_genotyped = sum(genotyped)) %>%
  dplyr::mutate(prop_genotyped = n_genotyped / n)
```

Which mutations had the top VAFs and what were their depths?

```{r}
muts_and_snps %>%
  dplyr::arrange(-mut_vaf) %>%
  dplyr::filter(mut_vaf > 0) %>%
  dplyr::select(source, sample_id, mut_id, mut_vaf, mut_depth, total_depth) %>%
  head(15)
```

How many mutations were shared between multiple cells?

```{r}
muts_and_snps %>%
  dplyr::filter(mut_depth > 0) %>%
  dplyr::group_by(mut_id, source) %>%
  dplyr::summarise(n_samples_w_mut = dplyr::n_distinct(sample_id),
                   mean_mut_vaf = mean(mut_vaf),
                   sample_ids = sample_id %>% unique() %>% sort() %>% paste(collapse = ",")) %>%
  dplyr::ungroup() %>%
  dplyr::count(source, n_samples_w_mut) %>%
  {split(., .$source)}
```

We can look at the distribution of VAFs.

```{r}
# distibution of VAFs
muts_and_snps %>%
  ggplot(aes(x = mut_vaf)) +
  geom_histogram(bins = 100) +
  ggtitle("Distribution of VAFs in PTA samples") +
  facet_grid(source ~ ., scales = "free_y")
```

We can look at the distribution of VAFs among mutations genotyped at least once.

```{r}
# distibution of non-zero VAFs
muts_and_snps %>%
  dplyr::filter(mut_depth > 0) %>%
  ggplot(aes(x = mut_vaf)) +
  geom_histogram(bins = 100) +
  ggtitle("Distribution of non-zero VAFs in PTA samples") +
  facet_grid(source ~ ., scales = "free_y")
```

We can look at the distribution of VAFs among mutations with at least 2 
supporting reads.

```{r}
seq(1, 5) %>%
  purrr::map(function(min_depth) {
    p_dat <-
      muts_and_snps %>%
      dplyr::filter(mut_depth >= min_depth, source == "nanoseq_mutations")
    p_dat %>%
      ggplot(aes(x = mut_vaf)) +
      geom_histogram(bins = 100) +
      ggtitle(paste0("Distribution of VAFs in PTA samples, min ",
                    min_depth, " mutant reads (",
                    dplyr::n_distinct(p_dat$mut_id), " mutations)")) +
      facet_grid(source ~ ., scales = "free_y") +
      scale_x_continuous(limits = c(0, 1))
  }) %>%
  patchwork::wrap_plots(ncol = 1)
```

We can look at the distribution of VAFs among mutations called more than once in 
Nanoseq. We expect these mutations to be in more cells and therefore to have
a higher genotyping efficiency.

```{r}
# distribution of VAFs in mutations called more than once in nanoseq
muts_and_snps %>%
  dplyr::filter(source == "nanoseq_mutations") %>%
  dplyr::full_join(nanoseq_muts) %>%
  dplyr::filter(times_called > 1) %>%
  ggplot(aes(x = mut_vaf)) +
  geom_histogram() +
  ggtitle("Distribution of VAFs in mutations called more than once in Nanoseq")
```

```{r}
# distribution of VAFs in mutations called more than once in nanoseq
muts_and_snps %>%
  dplyr::filter(source == "nanoseq_mutations", mut_depth > 0) %>%
  dplyr::full_join(nanoseq_muts) %>%
  dplyr::filter(times_called > 1) %>%
  ggplot(aes(x = mut_vaf)) +
  geom_histogram(bins = 100) +
  ggtitle("Distribution of non-zero VAFs in mutations called more than once in Nanoseq")
```

```{r}
# distribution of VAFs in mutations called more than once in nanoseq
muts_and_snps %>%
  dplyr::filter(source == "nanoseq_mutations") %>%
  dplyr::full_join(nanoseq_muts) %>%
  dplyr::mutate(times_called = ifelse(is.na(times_called), 0, times_called)) %>%
  ggplot(aes(x = times_called, y = mut_vaf, group = times_called)) +
  geom_point() +
  ggtitle("VAFs in PTA versus times called in Nanoseq")
```

