---
title: "ResolveOME celltype prediction"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: true
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
---

```{r setup, include = F, message = F, warning = F, echo = F}
# rmarkdown::render('src/08_celltype_prediction.Rmd', output_file = 'celltype_prediction.html', output_dir = 'out/analysis/')

# chunk options
knitr::opts_knit$set(root.dir = "../")
knitr::opts_chunk$set(warning = FALSE,
                      dpi = 300,
                      message = FALSE)

# libraries
library(magrittr)
library(ggplot2)
library(patchwork)
library(dplyr)

# celltype palette
pal <- c("B" = "#2f9e77", "mature B" = "#55c552",
         "non-mature B" = "#397037", "non-lymphocyte" = "#d95f02",
         "NK" = "#e72a8a", "Myeloid" = "#984ea3",
         "non-haematopoietic" = "#ff7f00", "T" = "#377eb8",
         "alpha-beta T cell" = "#377eb8", "B cell" = "#2f9e77",
         "mature B cell" = "#55c552", "not lymphocyte" = "#d95f02",
         "not mature B cell" = "#397037")
```

Here, I will synthesise the data from...

 - chromosomal dropout
 - doublet detection
 - celltype predictions from VDJ recombination
 - celltype predictions from BaseJumper bj-expression (SingleR + HPCA)
 - T and B cell fraction predictions from ImmuneLENS

...to get a better understanding of the celltypes present in the ResolveOME
dataset.

First, we load the samplesheet with all metadata for the cells.

```{r load_ss}
# load samplesheet
ss <-
  readr::read_csv("data/resolveome/samplesheet_local.csv") %>%
  mutate(run_id = as.character(run_id)) %>%
  filter(plate != 10)
ss_exp <-
  ss %>%
  select(id, cell_id, seq_type) %>%
  group_by(id, cell_id) %>%
  tidyr::expand_grid(name = c("celltype_SHM", "celltype_VDJ_recomb", "celltype_bj_expression")) %>%
  ungroup()
```

Next, we load the data from the manual inspection of the cells. This includes
whether we suspect them to be doublets or to have chromosomal dropout (based on
inspection of the BAF plot of heterozygous SNPs), celltype predictions based on
the inspection of VDJ recombination, and celltype predictions based on the 
presence or absence of somatic hypermutation.

```{r cts_man}
cts <- list()
cts$man <-
  "data/manual_inspection/2024-12-20_PD63118_PTA_BAF_LoH_CellType_Mut_Summary.tsv" %>%
  readr::read_tsv() %>%
  mutate(
    celltype_VDJ_SHM = case_when(
      celltype_VDJ_recomb == "B cell" & celltype_SHM == "not mature B cell" ~ "non-mature B",
      celltype_VDJ_recomb == "B cell" & celltype_SHM == "mature B cell" ~ "mature B",
      celltype_VDJ_recomb == "alpha-beta T cell" ~ "T",
      celltype_VDJ_recomb == "not lymphocyte" ~ "non-lymphocyte",
      TRUE ~ celltype_VDJ_recomb) %>% factor(levels = c("mature B", "non-mature B", "T", "non-lymphocyte"))) %>%
  select(cell_id, suspected_doublet, chr_dropout, celltype_VDJ_SHM,
         celltype_VDJ_recomb, celltype_SHM)
```

Next, we load the celltype predictions from BaseJumper, which uses SingleR to
make predictions based on the Human Primary Cell Atlas (HPCA).

```{r cts_bj}
# get latest bj dir
bj_dir <-
  list.files("out/BaseJumper/bj-expression/all_cells/", pattern = "PD63118_",
             include.dirs = TRUE, full.names = TRUE) %>%
  sort() %>%
  tail(1)

# get basejumper celltype preds
cts$bj <-
  readr::read_tsv(paste0(bj_dir, "/tertiary_analyses/classification_cell_typing/df_cell_typing_summary_singler_hpca_gtex_tcga.tsv")) %>%
  left_join(ss, by = c("SampleId" = "id")) %>%
  select(cell_id, celltype_bj_expression = Progenitor, tcga_tissue = TGCA_Tissue)
```

Next, we load the ImmuneLENS scores.

```{r cts_il}
cts$il <-
  "out/immunelens/scores.tsv" %>%
  readr::read_tsv() %>%
  left_join(ss) %>%
  transmute(cell_id, name = paste0("cell_frac_", gene), cell_fraction) %>%
  tidyr::pivot_wider(id_cols = "cell_id", values_from = "cell_fraction")
```

Now, we combine all of these lines of evidence for celltype in each cell. We 
subset this to only 'clean' cells as well - ie cells with no evidence of 
chromosomal dropout or of being a doublet.

```{r comb_preds}
# combine preds
cts <- Reduce(full_join, cts)

# get only 'clean' cells
cts_clean <-
  cts %>%
  filter(!suspected_doublet, !chr_dropout,
         !is.na(suspected_doublet), !is.na(chr_dropout))

# update samplesheet to only 'clean' cells
ss_clean <-
  ss %>%
  filter(cell_id %in% cts_clean$cell_id)
```

## VDJ + SHM predictions

```{r plot_vdj_shm}
plots <-
  cts_clean %>%
  arrange(celltype_VDJ_SHM) %>%
  mutate(cell_id = forcats::fct_inorder(cell_id)) %>%
  tidyr::pivot_longer(cols = c("celltype_VDJ_recomb", "celltype_SHM",
                               "celltype_VDJ_SHM"), values_to = "celltype") %>%
  {split(., .$name)} %>%
  purrr::map(function(p_dat) {
    p_dat %>%
      ggplot(aes(x = cell_id, y = name, fill = celltype)) +
      geom_tile() +
      scale_fill_manual(values = pal) +
      scale_y_discrete(expand = c(0, 0)) +
      guides(x = guide_axis(angle = -90)) +
      theme(axis.text.x = element_blank(), axis.title.x = element_blank(),
            axis.ticks.x = element_blank(), axis.title.y = element_blank())
  })
plots$celltype_VDJ_recomb <-
  plots$celltype_VDJ_recomb + theme(axis.text.x = element_text()) + scale_x_discrete(position = "top")
plots$celltype_VDJ_recomb / plots$celltype_SHM / plots$celltype_VDJ_SHM

cts_clean %>%
  count(celltype_VDJ_SHM) %>%
  ggplot(aes(x = reorder(celltype_VDJ_SHM, -n), y = n, fill = celltype_VDJ_SHM)) +
  geom_col() +
  geom_text(aes(label = n), vjust = -0.5) +
  guides(x = guide_axis(angle = -90)) +
  theme_classic() +
  scale_fill_manual(values = pal) +
  labs(x = "VDJ recombination celltype") +
  theme(legend.position = "none")
```

## BaseJumper bj-expression predictions

```{r plot_bj}
# plot celltype counts
cts %>%
  bind_rows(filter(cts, cell_id %in% ss_clean$cell_id) %>% mutate(group = "clean")) %>%
  mutate(group = ifelse(is.na(group), "full", group)) %>%
  count(celltype_bj_expression, group) %>%
  ggplot(aes(x = reorder(celltype_bj_expression, -n), y = n, fill = n)) +
  geom_col() +
  geom_text(aes(label = n), vjust = -0.5) +
  guides(x = guide_axis(angle = -90)) +
  theme_classic() +
  labs(x = "SingleR HPCA celltype prediction") +
  theme(legend.position = "none") +
  facet_grid(group ~ .)

# plot tcga tissue counts
cts %>%
  bind_rows(filter(cts, cell_id %in% ss_clean$cell_id) %>% mutate(group = "clean")) %>%
  mutate(group = ifelse(is.na(group), "full", group)) %>%
  count(tcga_tissue, group) %>%
  ggplot(aes(x = reorder(tcga_tissue, -n), y = n, fill = n)) +
  geom_col() +
  geom_text(aes(label = n), vjust = -0.5) +
  guides(x = guide_axis(angle = -90)) +
  theme_classic() +
  labs(x = "TCGA tissue") +
  theme(legend.position = "none") +
  facet_grid(group ~ .)

# plot clean celltype counts
cts_clean %>%
  add_count(celltype_bj_expression) %>%
  ggplot(aes(x = reorder(celltype_bj_expression, -n), fill = celltype_VDJ_SHM)) +
  geom_bar() +
  geom_text(aes(label = n, y = n), vjust = -0.5) +
  guides(x = guide_axis(angle = -90)) +
  theme_classic() +
  labs(x = "SingleR HPCA celltype prediction") +
  scale_fill_manual(values = pal) +
  ggtitle("BJ-Expression celltype vs VDJ recombination celltype")
```

## ImmuneLENS predictions

Now, we can plot the ImmuneLENS scores against the SHM / VDJ recombination-
predicted celltype.

```{r plot_vdj_vs_il}
cts %>%
  filter(!chr_dropout, !suspected_doublet) %>%
  tidyr::pivot_longer(cols = starts_with("cell_frac")) %>%
  ggplot(aes(x = celltype_VDJ_SHM, y = value)) +
  geom_boxplot(outlier.colour = NA) +
  geom_jitter(height = 0, aes(colour = suspected_doublet)) +
  facet_wrap(~ name) +
  ggtitle("ImmuneLENS scores vs VDJ recombination celltype")

cts_clean %>%
  tidyr::pivot_longer(cols = starts_with("cell_frac")) %>%
  filter(name %in% c("cell_frac_IGH", "cell_frac_TCRA")) %>%
  ggplot(aes(x = celltype_VDJ_SHM, y = value, colour = celltype_VDJ_SHM)) +
  geom_boxplot(outlier.colour = NA) +
  geom_jitter(height = 0, width = 0.1) +
  facet_grid(~ name) +
  ggtitle("ImmuneLENS scores vs VDJ recombination celltype") +
  scale_colour_manual(values = pal) +
  theme_bw() +
  theme(legend.position = "none")
```

Now, we can plot all celltype predictions next to each other.

```{r plot_cts}
p_vdj_shm <-
  cts_clean %>%
  ggplot(aes(x = cell_id, y = 0, fill = celltype_VDJ_SHM)) +
  geom_tile() +
  scale_y_discrete(expand = c(0, 0)) +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_blank()) +
  facet_grid(~ celltype_VDJ_SHM, scales = "free_x", space = "free_x") +
  scale_fill_manual(values = pal) +
  theme(strip.text = element_blank())

p_bje <-
  cts_clean %>%
  mutate(
    ct_VDJ_SHM = celltype_VDJ_SHM,
    celltype_bj_expression = case_when(
      celltype_bj_expression == "B_cell" ~ "B",
      celltype_bj_expression == "T_cells" ~ "T",
      celltype_bj_expression == "NK_cell" ~ "NK",
      celltype_bj_expression %in% c("Monocyte", "Platelets", "Neutrophils", "Macrophage") ~ "Myeloid",
      TRUE ~ "non-haematopoietic")) %>%
  ggplot(aes(x = cell_id, y = 0, fill = celltype_bj_expression)) +
  geom_tile() +
  scale_y_discrete(expand = c(0, 0)) +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_blank()) +
  facet_grid(~ celltype_VDJ_SHM, scales = "free_x", space = "free_x") +
  scale_fill_manual(values = pal) +
  theme(strip.text = element_blank())

p_ils <-
  cts_clean %>%
  tidyr::pivot_longer(cols = starts_with("cell_frac"), values_to = "cell fraction") %>%
  ggplot(aes(x = cell_id, y = name, fill = `cell fraction`)) +
  geom_tile() +
  scale_y_discrete(expand = c(0, 0)) +
  guides(x = guide_axis(angle = -90)) +
  theme(axis.title.y = element_blank()) +
  facet_grid(~ celltype_VDJ_SHM, scales = "free_x", space = "free_x") +
  theme(strip.text = element_blank())

p_vdj_shm / p_bje / p_ils + plot_layout(heights = c(1, 1, 4))
```

## Read and gene counts from RNA

```{r read_and_gene_counts}
# dirs
bje_dir <- "out/BaseJumper/bj-expression/PD63118_250225_102304/"

# load read counts
read_counts <-
  file.path(bje_dir, "read_counts/combined_read_counts.txt") %>%
  readr::read_csv() %>%
  transmute(id = sample_name, n = reads, name = "n reads")

# load gene counts
gene_counts <-
  file.path(bje_dir, "secondary_analyses/quantification_salmon/df_gene_counts_salmon.tsv") %>%
  readr::read_tsv() %>%
  mutate(id = gsub("salmon_outdir_", "", File))
gene_total_counts <-
  gene_counts %>%
  group_by(id) %>%
  summarise(n = n_distinct(gene_symbol_gene_id), name = "n genes")

# combine counts
counts <-
  bind_rows(read_counts, gene_total_counts) %>%
  left_join(ss)

# plot all runs
counts %>%
  # add reads per cell for ordering
  left_join(read_counts %>% select(id, reads = n)) %>%
  ggplot(aes(x = reorder(id, -reads), y = n, fill = run_id)) +
  geom_col() +
  guides(x = guide_axis(angle = -90)) +
  scale_y_continuous(expand = c(0, 0)) +
  labs(x = "") +
  facet_wrap(~ name, scales = "free_y", ncol = 1) +
  theme_bw()

# compare the metrics between runs
counts %>%
  ggplot(aes(x = run_id, y = n, fill = run_id)) +
  geom_boxplot(outlier.colour = NA) +
  geom_jitter(height = 0) +
  facet_wrap(~ name, scales = "free_y", ncol = 2) +
  theme(legend.position = "none") +
  theme_bw()

# compare the same cells between the two runs
counts %>%
  left_join(
    counts %>%
    filter(name == "n reads") %>%
    group_by(cell_id) %>%
    summarise(reads = max(n))) %>%
  filter(plate == 3) %>%
  ggplot(aes(x = reorder(cell_id, -reads), y = n, fill = run_id)) +
  geom_col(position = "dodge") +
  guides(x = guide_axis(angle = -90)) +
  scale_y_continuous(expand = c(0, 0)) +
  labs(x = "") +
  theme_bw() +
  facet_grid(name ~ ., scales = "free_y")
```

```{r ct_markers}
# get celltype curated markers from celltypist
ct_markers <-
  readr::read_tsv("../reference/celltypist/Basic_celltype_information.tsv") %>%
  janitor::clean_names() %>%
  tidyr::separate_longer_delim("curated_markers", delim = ", ") %>%
  rename(gene_symbol = curated_markers)

# convert to matrix for heatmap
p_dat <-
  gene_counts %>%
  # get only clean cells
  inner_join(ss_clean) %>%
  # get curated markers
  inner_join(ct_markers) %>%
  # aggregate by cell
  group_by(cell_id, gene_symbol) %>%
  mutate(count = sum(countsFromAbundanceNo, na.rm = TRUE))

# plot heatmap
p_dat %>%
  ggplot(aes(x = cell_id, y = gene_symbol, fill = count)) +
  geom_tile() +
  facet_grid() +
  guides(x = guide_axis(angle = -90)) +
  facet_grid(high_hierarchy_cell_types ~ ., scales = "free_y", space = "free_y") +
  theme_classic() +
  theme(panel.spacing = unit(0, "lines"),
        panel.background = element_rect(fill = "grey", color = NA)) +
  scale_fill_viridis()
```