---
title: "ResolveOME differential expression analysis"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: true
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
---

```{r setup, include = F, message = F, warning = F, echo = F}
# rmarkdown::render('src/09_differential_expression.Rmd', output_file = 'differential_expression.html', output_dir = 'out/analysis/')

# chunk options
knitr::opts_knit$set(root.dir = "../")
knitr::opts_chunk$set(warning = FALSE,
                      dpi = 300,
                      message = FALSE)

# libraries
library(magrittr)
library(ggplot2)
library(patchwork)
library(DESeq2)
library(dplyr)
library(tibble)
library(ggrepel)
library(pheatmap)
library(clusterProfiler)
library(org.Hs.eg.db)
```

Here, I will run a differential gene expression analysis between the WT cells
and the cells with *TNFRSF14*, *TET2*, and *CD274* mutations.

First, we load the samplesheet with all metadata for the cells.

```{r load_ss}
# load samplesheet
ss <-
  readr::read_csv("data/resolveome/samplesheet_local.csv") %>%
  mutate(run_id = as.character(run_id))

# load manual inspection data
man_insp <-
  readr::read_tsv("data/manual_inspection/2024-12-20_PD63118_PTA_BAF_LoH_CellType_Mut_Summary.tsv") %>%
  # clean cells
  filter(!chr_dropout | is.na(chr_dropout),
         !suspected_doublet | is.na(suspected_doublet)) %>%
  mutate(loh_1p = as.character(as.numeric(loh_1p))) %>%
  select(cell_id, plate, starts_with("celltype_"), TNFRSF14_mut, loh_1p) %>%
  distinct()
```

## Read and gene counts from RNA

```{r read_and_gene_counts}
# dirs
bje_dir <- "out/BaseJumper/bj-expression/rna/PD63118_run/"

# load read counts
read_counts <-
  file.path(bje_dir, "read_counts/combined_read_counts.txt") %>%
  readr::read_csv() %>%
  transmute(id = sample_name, n = reads, name = "n reads")

# load gene counts
gene_counts <-
  file.path(bje_dir, "secondary_analyses/quantification_salmon/df_gene_counts_salmon.tsv") %>%
  readr::read_tsv() %>%
  mutate(id = gsub("salmon_outdir_", "", File))
gene_total_counts <-
  gene_counts %>%
  group_by(id) %>%
  summarise(n = n_distinct(gene_symbol_gene_id), name = "n genes")

# combine counts
counts <-
  bind_rows(read_counts, gene_total_counts) %>%
  left_join(ss)

# plot all runs
counts %>%
  # add reads per cell for ordering
  left_join(read_counts %>% select(id, reads = n)) %>%
  ggplot(aes(x = reorder(id, -reads), y = n, fill = run_id)) +
  geom_col() +
  guides(x = guide_axis(angle = -90)) +
  scale_y_continuous(expand = c(0, 0)) +
  labs(x = "") +
  facet_grid(name ~ plate, scales = "free", space = "free_x") +
  theme_bw()
```

Plot n counts and n features per cell.

```{r}
# load expression matrix
expr <-
  file.path(bje_dir, "secondary_analyses/quantification_htseq/matrix_gene_counts_starhtseq.txt") %>%
  readr::read_tsv() %>%
  tidyr::separate_wider_delim(cols = "gene_id_gene_symbol", delim = "_",
                              names = c("gene_id", "gene_symbol")) %>%
  rename_with(~ stringr::str_replace(., "_rna_.*", ""))

# Compute nCounts and nFeatures for each cell (excluding gene_id and gene_symbol)
cell_stats <-
  expr %>%
  select(-gene_id, -gene_symbol) %>%
  summarise(
    `total counts` = colSums(.),          # Total counts per cell
    `total genes` = colSums(. > 0)     # Number of expressed genes per cell
  ) %>%
  tidyr::pivot_longer(cols = everything(), names_to = "metric", values_to = "value") %>%
  mutate(cell_id = rep(colnames(expr)[-c(1, 2)], 2)) # Add cell labels

# Plot
cell_stats %>%
  inner_join(man_insp) %>%
  group_by(cell_id, plate) %>%
  mutate(arr = max(value)) %>%
  ungroup() %>%
  ggplot(aes(x = reorder(cell_id, -arr), y = value, fill = value)) +
  geom_bar(stat = "identity", position = "dodge") +
  theme_minimal() +
  guides(x = guide_axis(angle = -90)) +
  labs(x = "cell", y = "") +
  facet_grid(metric ~ plate, scales = "free", space = "free_x") +
  scale_y_continuous(expand = c(0, 0)) +
  theme(panel.grid.minor.x = element_blank(),
        panel.grid.major.x = element_blank(),
        legend.position = "none") +
  viridis::scale_fill_viridis()
```

## DGEA

Next, we load the cell-gene count expression matrices.

```{r expr_mat}
# list of compound heterozygotes
comp_hets <-
  c("plate10_wellA11",
    "plate10_wellC6",
    "plate3_wellB4",
    "plate10_wellE10",
    "plate3_wellF11",
    "plate10_wellF6",
    "plate10_wellC9",
    "plate10_wellF7")

# get cells and their tnfrsf14 status
cells <-
  man_insp %>%
  mutate(sample_id = cell_id,
         condition = factor(as.numeric(!is.na(TNFRSF14_mut)),
                            levels = c("0", "1")),
         comp_het = as.numeric(cell_id %in% comp_hets)) %>%
  filter(celltype_VDJ_recomb == "B cell" & celltype_SHM == "mature B cell")

expr_cells <-
  expr %>%
  select(gene_id, gene_symbol, any_of(cells$sample_id))

# Ensure sample IDs in `cells` are in the same order as columns in `expr`
expr_counts <-
  expr_cells %>%
  column_to_rownames(var = "gene_id") %>%  # Use gene names as rownames
  select(-gene_symbol)  # Remove gene_id if not needed

# Ensure order of sample IDs matches
cells <-
  cells %>%
  filter(sample_id %in% colnames(expr_counts)) %>%
  arrange(match(sample_id, colnames(expr_counts)))

# Convert to a DESeq2-compatible format
dds <- DESeqDataSetFromMatrix(
  countData = as.matrix(expr_counts),
  colData = cells,
  design = ~ condition)

# Set reference level for condition (WT as reference)
dds$condition <- relevel(dds$condition, ref = "0")

# Run DESeq2 normalization and differential expression analysis
dds <- DESeq(dds)

# To visualize normalized counts for a specific gene
normalised_counts <- counts(dds, normalized = TRUE)

# Extract results
res <- results(dds, contrast = c("condition", "1", "0"))

# Order results by adjusted p-value (FDR)
res <- res[order(res$padj), ]

# Convert to a tibble for easier viewing
res_tbl <-
  as_tibble(res, rownames = "gene") %>%
  left_join(expr %>% select(gene = gene_id, gene_symbol)) %>%
  mutate(direction = ifelse(log2FoldChange > 0, "upregulated", "downregulated"))

# View significant genes (FDR < 0.05)
significant_genes <-
  res_tbl %>%
  filter(padj < 0.05)
print(significant_genes)
```

```{r volcano, fig.height = 10, fig.width = 10}
# plot volcano
labs <-
  res_tbl %>%
  group_by(direction) %>%
  top_n(10, wt = -log10(padj))
res_tbl %>%
  ggplot(aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.05), alpha = 0.6) +
  ggrepel::geom_text_repel(data = labs, aes(label = gene_symbol), size = 3,
                           vjust = -0.5) +
  scale_color_manual(values = c("black", "red")) +
  theme_minimal() +
  labs(title = "DGEA, TNFRSF14 mutant vs WT mature B cells",
       x = "Log2 Fold Change",
       y = "-Log10 Adjusted P-Value")
```

```{r heatmap, fig.height = 12, fig.width = 10}
# plot heatmap
# Get top 50 differentially expressed genes
top_genes <-
  significant_genes %>%
  group_by(direction) %>%
  top_n(50, wt = -log10(padj)) %>%
  {split(., .$direction)}

# function: plot heatmap
plot_heatmap <- function(top_genes_i, counts_i, p_title = "Top 50 DGEs",
                         annotations = c()) {
  # Extract their expression data
  expr_top <- expr_counts[top_genes_i$gene, arrange(cells, condition, comp_het)$sample_id]
  rownames(expr_top) <- top_genes_i$gene_symbol

  # Create a column annotation for TNFRSF14 status
  annotation_col <-
    cells %>%
    transmute(condition = factor(condition), celltype_SHM, sample_id, loh_1p, comp_het) %>%
    tibble::column_to_rownames(var = "sample_id")

  # Define color scheme
  ann_colors <- list(condition = c("0" = "blue", "1" = "red"))

  # Scale and visualize heatmap
  pheatmap(log2(expr_top + 1),
           scale = "row", cluster_cols = FALSE,
           clustering_distance_rows = "correlation",
           annotation_col = annotation_col,
           annotation_colors = ann_colors,
           main = p_title)
}

# plot up/downregulated genes
plot_heatmap(top_genes$upregulated, expr_counts, p_title = "Top 50 upregulated DGEs")
plot_heatmap(top_genes$downregulated, expr_counts, p_title = "Top 50 downregulated DGEs")
```

### NFKB pathway genes

```{r}
# expression of genes in the nfkb pathway
# List of NF-κB pathway genes from chatGPT
nfkb_genes <- c(
  "RELA", "RELB", "NFKB1", "NFKB2", "REL",
  "NFKBIA", "NFKBIB", "NFKBIE", "BCL3",
  "CHUK", "IKBKB", "IKBKG",
  "MAP3K7", "TRAF2", "TRAF5", "TRAF6", "RIPK1", "RIPK2", "TANK", "TNFAIP3",
  "TNFRSF1A", "TNFRSF1B", "CD40", "TNFRSF5", "TNFRSF14", "LTBR",
  "TLR2", "TLR4", "TLR9",
  "TNF", "IL1A", "IL1B", "IL6", "IL8", "CCL2", "CCL5"
)

# # nfkb genes from Biocarta Pathways dataset
# nfkb_genes <-
#   c("CHUK", "FADD", "IKBKB", "IKBKG", "IL1A", "IL1R1", "IRAK1", "MAP3K1",
#     "MAP3K7", "MAP4K4", "MYD88", "NFKB1", "NFKBIA", "RELA", "RIPK1", "TAB1",
#     "TLR4", "TNF", "TNFRSF1A", "TRADD")

# Select NF-κB genes from DESeq2 results
nfkb_res <-
  res_tbl %>%
  filter(gene_symbol %in% nfkb_genes) %>%
  arrange(padj)

# Print results
print(nfkb_res)
```

```{r nfkb_volcano, fig.height = 10, fig.width = 10}
res_tbl %>%
  ggplot(aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.05), alpha = 0.6) +
  scale_color_manual(values = c("black", "red")) +
  geom_label_repel(data = nfkb_res, aes(label = gene_symbol)) + 
  theme_minimal() +
  labs(title = "NF-κB Pathway Gene Expression",
       x = "Log2 Fold Change",
       y = "-Log10 Adjusted P-Value")
```

```{r nfkb_heatmap, fig.height = 10, fig.width = 10}
# Get expression matrix for NF-κB genes
expr_nfkb <- expr_counts[nfkb_res$gene, arrange(cells, condition)$sample_id]
rownames(expr_nfkb) <- nfkb_res$gene_symbol

# Annotate samples by TNFRSF14 status
annotation_col <- data.frame(condition = ifelse(cells$condition == "1", "mut", "WT"))
rownames(annotation_col) <- cells$sample_id
pheatmap(log2(expr_nfkb + 1),
         scale = "row",
         cluster_rows = F, cluster_cols = F,
         #clustering_distance_rows = "correlation",
         annotation_col = annotation_col,
         main = "NF-κB Pathway Gene Expression")
```

## Pathway enrichment analysis

Run pathway enrichment analysis on the cells.

```{r path_enrich}
# function: run GO and KEGG pathway enrichment on a list of genes
run_pathway_enrichment <- function(genes) {
  # map Ensembl to Entrez IDs for enrichment
  sig_entrez <-
    genes %>%
    tools::file_path_sans_ext() %>%
    bitr(fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = org.Hs.eg.db) %>%
    left_join(
      significant_genes %>% transmute(ENSEMBL = tools::file_path_sans_ext(gene),
                                      gene_symbol = gene_symbol))

  # run GO enrichment
  ego <- enrichGO(gene = sig_entrez$ENTREZID,
                  OrgDb = org.Hs.eg.db,
                  keyType = "ENTREZID",
                  ont = "BP",
                  pAdjustMethod = "BH",
                  pvalueCutoff = 0.05,
                  qvalueCutoff = 0.1)

  # run KEGG enrichment
  ekegg <- enrichKEGG(gene = sig_entrez$ENTREZID,
                      organism = "hsa",
                      pvalueCutoff = 0.05)

  # return
  list(ego = ego, ekegg = ekegg)
}

# run for upreg and downreg
pe_upreg <-
  significant_genes %>%
  filter(log2FoldChange > 1) %>%
  pull(gene) %>%
  run_pathway_enrichment()
pe_downreg <-
  significant_genes %>%
  filter(log2FoldChange < -1) %>%
  pull(gene) %>%
  run_pathway_enrichment()

# visualize top results
barplot(ego, showCategory = 10, title = "GO Biological Process Enrichment")
barplot(ekegg, showCategory = 10, title = "KEGG Pathway Enrichment")
```

```{r go_genes}
# upreg genes involved from GO output
pe_upreg$ego@result %>%
  dplyr::rename(ENTREZID = geneID) %>%
  filter(p.adjust < 0.05) %>%
  tidyr::separate_longer_delim(ENTREZID, delim = "/") %>%
  left_join(sig_entrez) %>%
  group_by(Description, GeneRatio, FoldEnrichment, p.adjust) %>%
  summarise(genes = paste(gene_symbol, collapse = ", "))

# downreg genes involved from GO output
pe_downreg$ego@result %>%
  dplyr::rename(ENTREZID = geneID) %>%
  filter(p.adjust < 0.05) %>%
  tidyr::separate_longer_delim(ENTREZID, delim = "/") %>%
  left_join(sig_entrez) %>%
  group_by(Description, GeneRatio, FoldEnrichment, p.adjust) %>%
  summarise(genes = paste(gene_symbol, collapse = ", "))
```

```{r kegg_genes}
# upreg genes involved from KEGG output
pe_upreg$ekegg@result %>%
  dplyr::rename(ENTREZID = geneID) %>%
  filter(p.adjust < 0.05) %>%
  tidyr::separate_longer_delim(ENTREZID, delim = "/") %>%
  left_join(sig_entrez) %>%
  group_by(Description, GeneRatio, FoldEnrichment, p.adjust) %>%
  summarise(genes = paste(gene_symbol, collapse = ", "))

# downreg genes involved from KEGG output
pe_downreg$ekegg@result %>%
  dplyr::rename(ENTREZID = geneID) %>%
  filter(p.adjust < 0.05) %>%
  tidyr::separate_longer_delim(ENTREZID, delim = "/") %>%
  left_join(sig_entrez) %>%
  group_by(Description, GeneRatio, FoldEnrichment, p.adjust) %>%
  summarise(genes = paste(gene_symbol, collapse = ", "))
```

## PCA

Run PCA.

```{r pca, fig.height = 10, fig.width = 10}
# Normalize the data using DESeq2
dds <- DESeqDataSetFromMatrix(countData = expr_counts,
                              colData = DataFrame(condition=rep("Condition1", ncol(expr_counts))), 
                              design = ~ 1) # Adjust 'design' as necessary

# Perform DESeq normalization
dds <- DESeq(dds)
norm_counts <- counts(dds, normalized = TRUE)

# Calculate variance for each gene and select top 500 genes with the highest variance
variances <- apply(norm_counts, 1, var)
top_500_genes <- order(variances, decreasing = TRUE)[1:500]

# Subset the data for the top 500 genes
top_500_counts <- norm_counts[top_500_genes, ]

# Perform PCA
pca_result <- prcomp(t(top_500_counts), scale. = TRUE)

# Create a data frame for the PCA results
pca_df <- data.frame(PC1 = pca_result$x[, 1],
                     PC2 = pca_result$x[, 2],
                     Cell = colnames(top_500_counts)) %>%
  left_join(cells, by = c("Cell" = "sample_id"))

# Plot the PCA results
pca_df %>%
  ggplot(aes(x = PC1, y = PC2, label = Cell)) +
  geom_point(aes(colour = condition)) +
  geom_text_repel(aes(label = Cell)) +
  theme_minimal() +
  ggtitle("PCA of Top 500 Genes") +
  xlab("Principal Component 1") +
  ylab("Principal Component 2")
```
