---
title: "Shared mutations from the bj-somatic-variantcalling output"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: true
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
---

```{r setup, include = F}
# rmarkdown::render('src/09_somatic_mutations.Rmd', output_file = 'somatic_mutations.html', output_dir = 'out/analysis/')

# chunk options
knitr::opts_knit$set(root.dir = "../")
knitr::opts_chunk$set(warning = FALSE,
                      dpi = 600,
                      message = FALSE)

# libraries
library(magrittr)
library(ggplot2)
library(dplyr)
```

We load the samplesheet.

```{r load_data}
ss <- readr::read_csv("out/BaseJumper/bj-somatic-variantcalling/dna/samplesheet.csv")
metadata <- readr::read_csv("data/resolveome/samplesheet_local.csv")
```

We load the annotated variants.

```{r load_vars}
# get the latest run
out_dir <-
  list.files("out/BaseJumper/bj-somatic-variantcalling/dna/",
             pattern = "PD63118_", include.dirs = TRUE, full.names = TRUE) %>%
  sort() %>%
  tail(1)

# increase buffer size
Sys.setenv("VROOM_CONNECTION_SIZE" = 5000000)

vafs <-
  ss$biosampleName %>%
  purrr::set_names() %>%
  purrr::map(function(id) {
    print(id)
    vcf_file <-
      paste0(out_dir, "/secondary_analyses/variant_calls_dnascope/",
             id, "_dnascope.vcf.gz")
    if (file.exists(vcf_file)) {
      vcf <-
        readr::read_tsv(gzfile(vcf_file), comment = "##") %>%
        {split(., .$FORMAT)} %>%
        purrr::map(function(vcf_i) {
          format_i <- stringr::str_split_1(unique(vcf_i$FORMAT), ":")
          vcf_i %>%
            mutate(gt = get(id)) %>%
            tidyr::separate_wider_delim("gt", delim = ":", names = format_i) %>%
            janitor::clean_names() %>%
            mutate(chr = number_chrom, total_depth = dp,
                          allele = paste(ref, alt, sep = ","),
                          alt = gsub(",.*", "", alt)) %>%
            tidyr::separate_longer_delim(cols = c("allele", "ad"), delim = ",") %>%
            select(chr, pos, ref, alt, allele, total_depth, ad) %>%
            mutate(name = case_when(allele == ref ~ "ref_depth",
                                                  allele == alt ~ "alt_depth",
                                                  TRUE ~ "other_depth")) %>%
            filter(name != "other_depth") %>%
            tidyr::pivot_wider(
              names_from = "name", values_from = "ad",
              id_cols = c("chr", "pos", "ref", "alt", "total_depth")) %>%
            readr::type_convert() %>%
            mutate(alt_vaf = alt_depth / total_depth)
        }) %>%
        bind_rows()
    }
  }) %>%
  purrr::compact() %>%
  bind_rows(.id = "id")

# save vafs
vafs %>% readr::write_tsv("out/analysis/dnahyb_bj-dna_mut_vafs.tsv")
```

Plot the VAF distribution.

```{r plot_vaf}
vafs %>%
  ggplot(aes(x = alt_vaf)) +
  geom_histogram()
```

Plot a heatmap of the shared variants.

```{r plot_heatmap}
# look at shared muts
p_dat <-
  vafs %>%
  # remove muts with VAF < 0.3
  filter(alt_vaf >= 0.3) %>%
  mutate(mut_id = paste(chr, pos, ref, alt, sep = "-"),
                n_cells = n_distinct(id)) %>%
  add_count(mut_id, name = "n_cells_w_mut") %>%
  mutate(prop_cells_w_mut = n_cells_w_mut / n_cells) %>%
  filter(n_cells_w_mut > 1, prop_cells_w_mut < 0.3)

# reshape data for heatmap
heatmap_data <-
  p_dat %>%
  reshape2::dcast(mut_id + n_cells_w_mut ~ id, value.var = "alt_vaf")
rownames(heatmap_data) <- heatmap_data$mut_id
heatmap_matrix <- as.matrix(heatmap_data[, -c(1:3)]) # drop annotations
n_total_muts <- n_distinct(p_dat$mut_id)
n_total_cells <- n_distinct(p_dat$id)

# calculate the number of mutations per column (id)
annotation_col <-
  p_dat %>%
  count(id, name = "n_muts") %>%
  left_join(metadata %>% select(run_id, id)) %>%
  mutate(run_id = as.character(run_id)) %>%
  tibble::column_to_rownames("id")

# replace NAs with 0
heatmap_matrix[is.na(heatmap_matrix)] <- 0

# prepare annotations for columns
annotation_row <- data.frame(n = heatmap_data$n)
rownames(annotation_row) <- heatmap_data$mut_id

# plot
pheatmap::pheatmap(
  mat = heatmap_matrix,
  cluster_rows = TRUE,
  cluster_cols = TRUE,
  show_rownames = FALSE,
  annotation_row = annotation_row,
  annotation_col = annotation_col,
  color = colorRampPalette(c("white", "blue"))(50),
  main = paste("VAF heatmap\n",
               n_total_cells, "cells,", n_total_muts, "mutations")
)
```